<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Home · GenomicBreeding.jl</title><meta name="title" content="Home · GenomicBreeding.jl"/><meta property="og:title" content="Home · GenomicBreeding.jl"/><meta property="twitter:title" content="Home · GenomicBreeding.jl"/><meta name="description" content="Documentation for GenomicBreeding.jl."/><meta property="og:description" content="Documentation for GenomicBreeding.jl."/><meta property="twitter:description" content="Documentation for GenomicBreeding.jl."/><meta property="og:url" content="https://GenomicBreeding.github.io/GenomicBreeding.jl/"/><meta property="twitter:url" content="https://GenomicBreeding.github.io/GenomicBreeding.jl/"/><link rel="canonical" href="https://GenomicBreeding.github.io/GenomicBreeding.jl/"/><script data-outdated-warner src="assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="assets/documenter.js"></script><script src="search_index.js"></script><script src="siteinfo.js"></script><script src="../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/catppuccin-mocha.css" data-theme-name="catppuccin-mocha"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/catppuccin-macchiato.css" data-theme-name="catppuccin-macchiato"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/catppuccin-frappe.css" data-theme-name="catppuccin-frappe"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/catppuccin-latte.css" data-theme-name="catppuccin-latte"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href>GenomicBreeding.jl</a></span></div><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li class="is-active"><a class="tocitem" href>Home</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>Home</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Home</a></li></ul></nav><div class="docs-right"><a class="docs-navbar-link" href="https://github.com/GenomicBreeding/GenomicBreeding.jl" title="View the repository on GitHub"><span class="docs-icon fa-brands"></span><span class="docs-label is-hidden-touch">GitHub</span></a><a class="docs-navbar-link" href="https://github.com/GenomicBreeding/GenomicBreeding.jl/blob/main/docs/src/index.md" title="Edit source on GitHub"><span class="docs-icon fa-solid"></span></a><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><h1 id="GenomicBreeding"><a class="docs-heading-anchor" href="#GenomicBreeding">GenomicBreeding</a><a id="GenomicBreeding-1"></a><a class="docs-heading-anchor-permalink" href="#GenomicBreeding" title="Permalink"></a></h1><p>Documentation for <a href="https://github.com/jeffersonfparil/GenomicBreeding.jl">GenomicBreeding</a>.</p><ul><li><a href="#GenomicBreeding.GBInput"><code>GenomicBreeding.GBInput</code></a></li><li><a href="#Base.:==-Tuple{GBInput, GBInput}"><code>Base.:==</code></a></li><li><a href="#Base.hash-Tuple{GBInput, UInt64}"><code>Base.hash</code></a></li><li><a href="#GBCore.checkdims-Tuple{GBInput}"><code>GBCore.checkdims</code></a></li><li><a href="#GenomicBreeding.assess-Tuple{GBInput}"><code>GenomicBreeding.assess</code></a></li><li><a href="#GenomicBreeding.clone-Tuple{GBInput}"><code>GenomicBreeding.clone</code></a></li><li><a href="#GenomicBreeding.extracteffects-Tuple{GBInput}"><code>GenomicBreeding.extracteffects</code></a></li><li><a href="#GenomicBreeding.load-Tuple{GBInput}"><code>GenomicBreeding.load</code></a></li><li><a href="#GenomicBreeding.plot-Tuple{}"><code>GenomicBreeding.plot</code></a></li><li><a href="#GenomicBreeding.preparegwasinputs-Tuple{GBInput}"><code>GenomicBreeding.preparegwasinputs</code></a></li><li><a href="#GenomicBreeding.prepareinputs-Tuple{GBInput}"><code>GenomicBreeding.prepareinputs</code></a></li><li><a href="#GenomicBreeding.submitslurmarrayjobs-Tuple{}"><code>GenomicBreeding.submitslurmarrayjobs</code></a></li></ul><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicBreeding.GBInput" href="#GenomicBreeding.GBInput"><code>GenomicBreeding.GBInput</code></a> — <span class="docstring-category">Type</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">mutable struct GBInput
    fname_geno::String
    fname_pheno::String
    bulk_cv::Bool
    populations::Union{Nothing,Vector{String}}
    traits::Union{Nothing,Vector{String}}
    models::Any
    n_folds::Int64
    n_replications::Int64
    keep_all::Bool
    gwas_models::Any
    maf::Float64
    mtv::Float64
    n_iter::Int64
    n_burnin::Int64
    fname_out_prefix::String
    SLURM_job_name::String
    SLURM_account_name::String
    SLURM_partition_name::String
    SLURM_nodes_per_array_job::Int64
    SLURM_tasks_per_node::Int64
    SLURM_cpus_per_task::Int64
    SLURM_mem_G::Int64
    SLURM_time_limit_dd_hhmmss::String
    SLURM_max_array_jobs_running::Int64
    SLURM_module_load_R_version_name::String
    SLURM_module_load_BLAS_version_name::String
    verbose::Bool
end</code></pre><p>Input struct (belongs to GBCore.AbstractGB type)</p><ul><li><code>fname_geno</code>: genotype file (see file format guide: <strong>TODO:</strong> {URL HERE})</li><li><code>fname_pheno</code>: phenotype file (see file format guide: <strong>TODO:</strong> {URL HERE})</li><li><code>bulk_cv</code>: perform cross-validation across all populations, i.e. disregard population grouping (Default = false)</li><li><code>populations</code>: include only these populations (Default = nothing which means include all populations)</li><li><code>traits</code>: include only these traits (Default = nothing which means include all traits)</li><li><code>models</code>: genomic prediction model functions (Default = [ridge, bayesa]; see models list: <strong>TODO:</strong> {URL HERE})</li><li><code>n_folds</code>: number of k partitions for k-fold cross-validation (Default = 5)</li><li><code>n_replications</code>: number of replications for repeated k-fold cross-validation (Default = 5)</li><li><code>gwas_models</code>: models to use if performning genome-wide association study (Default = [gwasols, gwaslmm])</li><li><code>keep_all</code>: keep all entries upon merging genomes and phenomes potentially resulting in sparsities in both structs? (Default = false)</li><li><code>maf</code>: minimum allele frequency (Default = 0.05)</li><li><code>mtv</code>: minimum trait variance (Default = 1e-7)</li><li><code>n_iter</code>: number of Bayesian model fitting MCMC/HMC iteration (Default = 1_500)</li><li><code>n_burnin</code>: number of initial Bayesian model fitting MCMC/HMC iterations to be excluded from the posterior distribution (Default = 500)</li><li><code>fname_out_prefix</code>: prefix of the output files which may include directory names (Default = &quot;&quot; which translates to <code>./GBOuput/output-&lt;yyyymmddHHMMSS&gt;-&lt;3_digit_random_number&gt;-</code>)</li><li><code>SLURM_job_name</code>: name of the Slurm job array (Default = &quot;&quot;)</li><li><code>SLURM_account_name</code>: Slurm account name (Default = &quot;&quot;)</li><li><code>SLURM_partition_name</code>: Slurm paritition to use (Default = &quot;&quot;)</li><li><code>SLURM_nodes_per_array_job</code>: number of nodes per array job (Default = 1)</li><li><code>SLURM_tasks_per_node</code>: number of tasks per node (Default = 1)</li><li><code>SLURM_cpus_per_task</code>: number of CPU cores per task (Default = 1)</li><li><code>SLURM_mem_G</code>: maximum memroy requested in gigabytes (Default = 1)</li><li><code>SLURM_time_limit_dd_hhmmss</code>: maximum computation time requested in the follwowing format: &quot;dd-hh:mm:ss&quot; (Default = &quot;00-01:00:00&quot;)</li><li><code>SLURM_max_array_jobs_running</code>: maximum number of array jobs which can run simultaneously (Default = 20)</li><li><code>SLURM_module_load_R_version_name</code>: name of the R statistical language module and version to be used to call R::BGLR (Default = &quot;R&quot;)</li><li><code>SLURM_module_load_BLAS_version_name</code>: name of the BLAS module and version to be used to use with R::BGLR. This is non-critical as there is usually a system default to fallback to. (Default = &quot;&quot;)</li><li><code>verbose</code>: show messages (Default = true)</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GenomicBreeding.jl/blob/b506265fa16bbb4f8063ecc4e42e789e09df144c/src/io.jl#L1-L61">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="Base.:==-Tuple{GBInput, GBInput}" href="#Base.:==-Tuple{GBInput, GBInput}"><code>Base.:==</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">Base.:(==)(x::GBInput, y::GBInput)::Bool</code></pre><p>Equality of GBInput structs using the hash function defined for GBInput structs.</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; input_1 = input = GBInput(fname_geno=&quot;geno1.jld2&quot;, fname_pheno=&quot;pheno1.jld2&quot;, fname_out_prefix=&quot;test1-&quot;, SLURM_job_name=&quot;slurmjob1&quot;);

julia&gt; input_2 = input = GBInput(fname_geno=&quot;geno1.jld2&quot;, fname_pheno=&quot;pheno1.jld2&quot;, fname_out_prefix=&quot;test1-&quot;, SLURM_job_name=&quot;slurmjob1&quot;);

julia&gt; input_3 = input = GBInput(fname_geno=&quot;geno2.jld2&quot;, fname_pheno=&quot;pheno2.jld2&quot;);

julia&gt; input_1 == input_2
true

julia&gt; input_1 == input_3
false</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GenomicBreeding.jl/blob/b506265fa16bbb4f8063ecc4e42e789e09df144c/src/io.jl#L187-L206">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="Base.hash-Tuple{GBInput, UInt64}" href="#Base.hash-Tuple{GBInput, UInt64}"><code>Base.hash</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">Base.hash(x::GBInput, h::UInt)::UInt</code></pre><p>Hash a GBInput struct using all its fields. We deliberately excluded the allele_frequencies, and mask for efficiency.</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; input = GBInput(fname_geno=&quot;&quot;, fname_pheno=&quot;&quot;);

julia&gt; typeof(hash(input))
UInt64</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GenomicBreeding.jl/blob/b506265fa16bbb4f8063ecc4e42e789e09df144c/src/io.jl#L163-L176">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GBCore.checkdims-Tuple{GBInput}" href="#GBCore.checkdims-Tuple{GBInput}"><code>GBCore.checkdims</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">checkdims(input::GBInput)::Bool</code></pre><p>Check dimension compatibility of the fields of the GBInput struct</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; input = GBInput(fname_geno=&quot;geno1.jld2&quot;, fname_pheno=&quot;pheno1.jld2&quot;);

julia&gt; checkdims(input)
true

julia&gt; input.models = nothing

julia&gt; checkdims(input)
false</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GenomicBreeding.jl/blob/b506265fa16bbb4f8063ecc4e42e789e09df144c/src/io.jl#L236-L253">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicBreeding.assess-Tuple{GBInput}" href="#GenomicBreeding.assess-Tuple{GBInput}"><code>GenomicBreeding.assess</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">assess(input::GBInput)::Tuple{Vector{String},Vector{String}}</code></pre><p>Assess genomic prediction accuracy via replicated k-fold cross-validation. Outputs are saved as JLD2 (each containing a CV struct per fold, replication, and trait) and possibly text file/s containing notes describing why some jobs failed.</p><p><strong>Example</strong></p><pre><code class="language-julia-repl hljs">julia&gt; genomes = GBCore.simulategenomes(n=300, l=1_000, verbose=false); genomes.populations = StatsBase.sample(string.(&quot;pop_&quot;, 1:3), length(genomes.entries), replace=true);

julia&gt; proportion_of_variance = fill(0.0, 9, 3); proportion_of_variance[1, :] .= 1.00; # 100% variance on the additive genetic effects

julia&gt; trials, _ = GBCore.simulatetrials(genomes=genomes, n_years=1, n_seasons=1, n_harvests=1, n_sites=1, n_replications=1, proportion_of_variance=proportion_of_variance, verbose=false);

julia&gt; phenomes = extractphenomes(trials);

julia&gt; fname_geno = try writedelimited(genomes, fname=&quot;test-geno.tsv&quot;); catch; rm(&quot;test-geno.tsv&quot;); writedelimited(genomes, fname=&quot;test-geno.tsv&quot;); end;

julia&gt; fname_pheno = try writedelimited(phenomes, fname=&quot;test-pheno.tsv&quot;); catch; rm(&quot;test-pheno.tsv&quot;); writedelimited(phenomes, fname=&quot;test-pheno.tsv&quot;); end;

julia&gt; input = GBInput(fname_geno=fname_geno, fname_pheno=fname_pheno, populations=[&quot;pop_1&quot;, &quot;pop_3&quot;], traits=[&quot;trait_1&quot;], n_replications=2, n_folds=3, verbose=false);

julia&gt; fnames_cvs, fnames_notes = assess(input);

julia&gt; length(fnames_cvs) == 32, length(fnames_notes) == 0
(true, true)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GenomicBreeding.jl/blob/b506265fa16bbb4f8063ecc4e42e789e09df144c/src/genomic_prediction.jl#L1-L28">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicBreeding.clone-Tuple{GBInput}" href="#GenomicBreeding.clone-Tuple{GBInput}"><code>GenomicBreeding.clone</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">clone(x::GBInput)::GBInput</code></pre><p>Clone a GBInput object</p><p><strong>Example</strong></p><pre><code class="language-julia-repl hljs">julia&gt; input = GBInput(fname_geno=&quot;geno1.jld2&quot;, fname_pheno=&quot;pheno1.jld2&quot;);

julia&gt; copy_input = clone(input);

julia&gt; input == copy_input
true</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GenomicBreeding.jl/blob/b506265fa16bbb4f8063ecc4e42e789e09df144c/src/io.jl#L211-L225">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicBreeding.extracteffects-Tuple{GBInput}" href="#GenomicBreeding.extracteffects-Tuple{GBInput}"><code>GenomicBreeding.extracteffects</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">extracteffects(input::GBInput)::Vector{String}</code></pre><p>Extract allele effects by fitting the models without cross-validation. The output are filenames of tab-delimited files containing the allele effects per model, trait and population.</p><p><strong>Example</strong></p><pre><code class="language-julia-repl hljs">julia&gt; genomes = GBCore.simulategenomes(n=300, l=1_000, verbose=false); genomes.populations = StatsBase.sample(string.(&quot;pop_&quot;, 1:3), length(genomes.entries), replace=true);

julia&gt; proportion_of_variance = fill(0.0, 9, 3); proportion_of_variance[1, :] .= 1.00; # 100% variance on the additive genetic effects

julia&gt; trials, _ = GBCore.simulatetrials(genomes=genomes, n_years=1, n_seasons=1, n_harvests=1, n_sites=1, n_replications=1, proportion_of_variance=proportion_of_variance, verbose=false);

julia&gt; phenomes = extractphenomes(trials);

julia&gt; fname_geno = try writedelimited(genomes, fname=&quot;test-geno.tsv&quot;); catch; rm(&quot;test-geno.tsv&quot;); writedelimited(genomes, fname=&quot;test-geno.tsv&quot;); end;

julia&gt; fname_pheno = try writedelimited(phenomes, fname=&quot;test-pheno.tsv&quot;); catch; rm(&quot;test-pheno.tsv&quot;); writedelimited(phenomes, fname=&quot;test-pheno.tsv&quot;); end;

julia&gt; input = GBInput(fname_geno=fname_geno, fname_pheno=fname_pheno, populations=[&quot;pop_1&quot;, &quot;pop_3&quot;], traits=[&quot;trait_1&quot;], n_replications=2, n_folds=3, verbose=false);

julia&gt; fname_effects = extracteffects(input);

julia&gt; length(fname_effects) == 6
true</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GenomicBreeding.jl/blob/b506265fa16bbb4f8063ecc4e42e789e09df144c/src/genomic_prediction.jl#L109-L136">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicBreeding.load-Tuple{GBInput}" href="#GenomicBreeding.load-Tuple{GBInput}"><code>GenomicBreeding.load</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">load(input::GBInput)::Tuple{Genomes, Phenomes}</code></pre><p>Load, merge and filter genotype and phenotype data</p><p><strong>Example</strong></p><pre><code class="language-julia-repl hljs">julia&gt; genomes = GBCore.simulategenomes(n=300, verbose=false); genomes.populations = StatsBase.sample(string.(&quot;pop_&quot;, 1:3), length(genomes.entries), replace=true);

julia&gt; trials, _ = GBCore.simulatetrials(genomes=genomes, n_years=1, n_seasons=1, n_harvests=1, n_sites=1, n_replications=1, verbose=false);

julia&gt; phenomes = extractphenomes(trials);

julia&gt; fname_geno = try writedelimited(genomes, fname=&quot;test-geno.tsv&quot;); catch; rm(&quot;test-geno.tsv&quot;); writedelimited(genomes, fname=&quot;test-geno.tsv&quot;); end;
    
julia&gt; fname_pheno = try writedelimited(phenomes, fname=&quot;test-pheno.tsv&quot;); catch; rm(&quot;test-pheno.tsv&quot;); writedelimited(phenomes, fname=&quot;test-pheno.tsv&quot;); end;

julia&gt; input = GBInput(fname_geno=fname_geno, fname_pheno=fname_pheno, populations=[&quot;pop_1&quot;, &quot;pop_3&quot;], traits=[&quot;trait_1&quot;], verbose=false);

julia&gt; genomes, phenomes = load(input);

julia&gt; length(unique(genomes.populations)) == length(unique(phenomes.populations)) == 2
true

julia&gt; length(phenomes.traits) == 1
true

julia&gt; rm.([fname_geno, fname_pheno]);</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GenomicBreeding.jl/blob/b506265fa16bbb4f8063ecc4e42e789e09df144c/src/io.jl#L258-L287">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicBreeding.plot-Tuple{}" href="#GenomicBreeding.plot-Tuple{}"><code>GenomicBreeding.plot</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">plot(;
    input::GBInput,
    skip_genomes::Bool = false,
    skip_phenomes::Bool = false,
    format::String = &quot;svg&quot;,
    plot_size::Tuple{Int64,Int64} = (600, 450),
    overwrite::Bool = false,
)::String</code></pre><p>Plot genomes, phenomes, and CVs, if present.</p><p><strong>Example</strong></p><pre><code class="language-julia-repl hljs">julia&gt; genomes = GBCore.simulategenomes(n=300, l=1_000, verbose=false); genomes.populations = StatsBase.sample(string.(&quot;pop_&quot;, 1:3), length(genomes.entries), replace=true);

julia&gt; trials, _ = GBCore.simulatetrials(genomes=genomes, n_years=1, n_seasons=1, n_harvests=1, n_sites=1, n_replications=1, verbose=false);

julia&gt; phenomes = extractphenomes(trials);

julia&gt; fname_geno = try writedelimited(genomes, fname=&quot;test-geno.tsv&quot;); catch; rm(&quot;test-geno.tsv&quot;); writedelimited(genomes, fname=&quot;test-geno.tsv&quot;); end;

julia&gt; fname_pheno = try writedelimited(phenomes, fname=&quot;test-pheno.tsv&quot;); catch; rm(&quot;test-pheno.tsv&quot;); writedelimited(phenomes, fname=&quot;test-pheno.tsv&quot;); end;
    
julia&gt; input = GBInput(fname_geno=fname_geno, fname_pheno=fname_pheno, SLURM_cpus_per_task=6, SLURM_mem_G=5, fname_out_prefix=&quot;GBOutput/test-&quot;, verbose=false);

julia&gt; GenomicBreeding.plot(input=input, format=&quot;png&quot;, plot_size = (700, 525))
&quot;GBOutput/plots&quot;

julia&gt; GenomicBreeding.plot(input=input, format=&quot;png&quot;, plot_size = (700, 525), overwrite=true, skip_genomes=true)
&quot;GBOutput/plots&quot;</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GenomicBreeding.jl/blob/b506265fa16bbb4f8063ecc4e42e789e09df144c/src/plot.jl#L1-L33">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicBreeding.preparegwasinputs-Tuple{GBInput}" href="#GenomicBreeding.preparegwasinputs-Tuple{GBInput}"><code>GenomicBreeding.preparegwasinputs</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">preparegwasinputs(input::GBInput)::Vector{GBInput}</code></pre><p>Prepare GBInputs for Slurm array jobs</p><p><strong>Example</strong></p><pre><code class="language-julia-repl hljs">julia&gt; genomes = GBCore.simulategenomes(n=300, verbose=false); genomes.populations = StatsBase.sample(string.(&quot;pop_&quot;, 1:3), length(genomes.entries), replace=true);

julia&gt; trials, _ = GBCore.simulatetrials(genomes=genomes, n_years=1, n_seasons=1, n_harvests=1, n_sites=1, n_replications=1, verbose=false);

julia&gt; phenomes = extractphenomes(trials);

julia&gt; fname_geno = try writedelimited(genomes, fname=&quot;test-geno.tsv&quot;); catch; rm(&quot;test-geno.tsv&quot;); writedelimited(genomes, fname=&quot;test-geno.tsv&quot;); end;
    
julia&gt; fname_pheno = try writedelimited(phenomes, fname=&quot;test-pheno.tsv&quot;); catch; rm(&quot;test-pheno.tsv&quot;); writedelimited(phenomes, fname=&quot;test-pheno.tsv&quot;); end;

julia&gt; input = GBInput(fname_geno=fname_geno, fname_pheno=fname_pheno, verbose=false);

julia&gt; inputs = preparegwasinputs(input);

julia&gt; length(inputs) == 30
true

julia&gt; rm.([fname_geno, fname_pheno]);</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GenomicBreeding.jl/blob/b506265fa16bbb4f8063ecc4e42e789e09df144c/src/io.jl#L598-L624">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicBreeding.prepareinputs-Tuple{GBInput}" href="#GenomicBreeding.prepareinputs-Tuple{GBInput}"><code>GenomicBreeding.prepareinputs</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">prepareinputs(input::GBInput)::Vector{GBInput}</code></pre><p>Prepare GBInputs for Slurm array jobs</p><p><strong>Example</strong></p><pre><code class="language-julia-repl hljs">julia&gt; genomes = GBCore.simulategenomes(n=300, verbose=false); genomes.populations = StatsBase.sample(string.(&quot;pop_&quot;, 1:3), length(genomes.entries), replace=true);

julia&gt; trials, _ = GBCore.simulatetrials(genomes=genomes, n_years=1, n_seasons=1, n_harvests=1, n_sites=1, n_replications=1, verbose=false);

julia&gt; phenomes = extractphenomes(trials);

julia&gt; fname_geno = try writedelimited(genomes, fname=&quot;test-geno.tsv&quot;); catch; rm(&quot;test-geno.tsv&quot;); writedelimited(genomes, fname=&quot;test-geno.tsv&quot;); end;
    
julia&gt; fname_pheno = try writedelimited(phenomes, fname=&quot;test-pheno.tsv&quot;); catch; rm(&quot;test-pheno.tsv&quot;); writedelimited(phenomes, fname=&quot;test-pheno.tsv&quot;); end;

julia&gt; input = GBInput(fname_geno=fname_geno, fname_pheno=fname_pheno, verbose=false);

julia&gt; inputs = prepareinputs(input);

julia&gt; length(inputs) == 30
true

julia&gt; rm.([fname_geno, fname_pheno]);</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GenomicBreeding.jl/blob/b506265fa16bbb4f8063ecc4e42e789e09df144c/src/io.jl#L519-L545">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicBreeding.submitslurmarrayjobs-Tuple{}" href="#GenomicBreeding.submitslurmarrayjobs-Tuple{}"><code>GenomicBreeding.submitslurmarrayjobs</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">submitslurmarrayjobs(; input::GBInput, analysis::Function)::String</code></pre><p>Assess genomic prediction accuracy via replicated k-fold cross-validation. Outputs are saved as JLD2 (each containing a CV struct per fold, replication, and trait) and possibly text file/s containing notes describing why some jobs failed. Note that you will be prompted to enter YES to proceed with Slurm job submission after showing you the job details to review and confirm.</p><p><strong>Example</strong></p><p>&lt;!– ```jldoctest; setup = :(using GBCore, GBIO, GenomicBreeding, StatsBase, DataFrames) –&gt;</p><pre><code class="nohighlight hljs">julia&gt; genomes = GBCore.simulategenomes(n=300, l=1_000, verbose=false); genomes.populations = StatsBase.sample(string.(&quot;pop_&quot;, 1:3), length(genomes.entries), replace=true);

julia&gt; trials, _ = GBCore.simulatetrials(genomes=genomes, n_years=1, n_seasons=1, n_harvests=1, n_sites=1, n_replications=1, verbose=false);

julia&gt; phenomes = extractphenomes(trials);

julia&gt; fname_geno = try writedelimited(genomes, fname=&quot;test-geno.tsv&quot;); catch; rm(&quot;test-geno.tsv&quot;); writedelimited(genomes, fname=&quot;test-geno.tsv&quot;); end;

julia&gt; fname_pheno = try writedelimited(phenomes, fname=&quot;test-pheno.tsv&quot;); catch; rm(&quot;test-pheno.tsv&quot;); writedelimited(phenomes, fname=&quot;test-pheno.tsv&quot;); end;

julia&gt; input = GBInput(fname_geno=fname_geno, fname_pheno=fname_pheno, SLURM_cpus_per_task=6, SLURM_mem_G=5, fname_out_prefix=&quot;GBOutput/test-&quot;, verbose=false);

julia&gt; outdir = submitslurmarrayjobs(input=input, analysis=assess)
GBOutput

julia&gt; run(`squeue`)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GenomicBreeding.jl/blob/b506265fa16bbb4f8063ecc4e42e789e09df144c/src/slurm.jl#L1-L28">source</a></section></article></article><nav class="docs-footer"><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="auto">Automatic (OS)</option><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option><option value="catppuccin-latte">catppuccin-latte</option><option value="catppuccin-frappe">catppuccin-frappe</option><option value="catppuccin-macchiato">catppuccin-macchiato</option><option value="catppuccin-mocha">catppuccin-mocha</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.8.1 on <span class="colophon-date" title="Thursday 13 February 2025 05:41">Thursday 13 February 2025</span>. Using Julia version 1.11.3.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
