<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>GenomicBreeding · GenomicBreeding.jl</title><meta name="title" content="GenomicBreeding · GenomicBreeding.jl"/><meta property="og:title" content="GenomicBreeding · GenomicBreeding.jl"/><meta property="twitter:title" content="GenomicBreeding · GenomicBreeding.jl"/><meta name="description" content="Documentation for GenomicBreeding.jl."/><meta property="og:description" content="Documentation for GenomicBreeding.jl."/><meta property="twitter:description" content="Documentation for GenomicBreeding.jl."/><meta property="og:url" content="https://jeffersonfparil.github.io/GenomicBreeding.jl/"/><meta property="twitter:url" content="https://jeffersonfparil.github.io/GenomicBreeding.jl/"/><link rel="canonical" href="https://jeffersonfparil.github.io/GenomicBreeding.jl/"/><script data-outdated-warner src="assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="assets/documenter.js"></script><script src="search_index.js"></script><script src="siteinfo.js"></script><script src="../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/catppuccin-mocha.css" data-theme-name="catppuccin-mocha"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/catppuccin-macchiato.css" data-theme-name="catppuccin-macchiato"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/catppuccin-frappe.css" data-theme-name="catppuccin-frappe"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/catppuccin-latte.css" data-theme-name="catppuccin-latte"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href>GenomicBreeding.jl</a></span></div><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li class="is-active"><a class="tocitem" href>GenomicBreeding</a></li><li><a class="tocitem" href="simulations/">Simulation</a></li><li><a class="tocitem" href="models/">Models</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>GenomicBreeding</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>GenomicBreeding</a></li></ul></nav><div class="docs-right"><a class="docs-navbar-link" href="https://github.com/jeffersonfparil/GenomicBreeding.jl" title="View the repository on GitHub"><span class="docs-icon fa-brands"></span><span class="docs-label is-hidden-touch">GitHub</span></a><a class="docs-navbar-link" href="https://github.com/jeffersonfparil/GenomicBreeding.jl/blob/main/docs/src/index.md" title="Edit source on GitHub"><span class="docs-icon fa-solid"></span></a><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><h1 id="GenomicBreeding"><a class="docs-heading-anchor" href="#GenomicBreeding">GenomicBreeding</a><a id="GenomicBreeding-1"></a><a class="docs-heading-anchor-permalink" href="#GenomicBreeding" title="Permalink"></a></h1><p>Documentation for <a href="https://github.com/jeffersonfparil/GenomicBreeding.jl">GenomicBreeding</a>.</p><ul><li><a href="#GenomicBreeding.Genomes"><code>GenomicBreeding.Genomes</code></a></li><li><a href="#GenomicBreeding.Phenomes"><code>GenomicBreeding.Phenomes</code></a></li><li><a href="#GenomicBreeding.SimulatedEffects"><code>GenomicBreeding.SimulatedEffects</code></a></li><li><a href="#GenomicBreeding.TEBV"><code>GenomicBreeding.TEBV</code></a></li><li><a href="#GenomicBreeding.Trials"><code>GenomicBreeding.Trials</code></a></li><li><a href="#Base.:==-Tuple{TEBV, TEBV}"><code>Base.:==</code></a></li><li><a href="#Base.:==-Tuple{Genomes, Genomes}"><code>Base.:==</code></a></li><li><a href="#Base.:==-Tuple{Trials, Trials}"><code>Base.:==</code></a></li><li><a href="#Base.:==-Tuple{Phenomes, Phenomes}"><code>Base.:==</code></a></li><li><a href="#Base.:==-Tuple{SimulatedEffects, SimulatedEffects}"><code>Base.:==</code></a></li><li><a href="#Base.hash-Tuple{Trials, UInt64}"><code>Base.hash</code></a></li><li><a href="#Base.hash-Tuple{Phenomes, UInt64}"><code>Base.hash</code></a></li><li><a href="#Base.hash-Tuple{TEBV, UInt64}"><code>Base.hash</code></a></li><li><a href="#Base.hash-Tuple{Genomes, UInt64}"><code>Base.hash</code></a></li><li><a href="#Base.hash-Tuple{SimulatedEffects, UInt64}"><code>Base.hash</code></a></li><li><a href="#Base.sum-Tuple{SimulatedEffects}"><code>Base.sum</code></a></li><li><a href="#GenomicBreeding.analyse-Tuple{Trials}"><code>GenomicBreeding.analyse</code></a></li><li><a href="#GenomicBreeding.checkdims-Tuple{Trials}"><code>GenomicBreeding.checkdims</code></a></li><li><a href="#GenomicBreeding.checkdims-Tuple{SimulatedEffects}"><code>GenomicBreeding.checkdims</code></a></li><li><a href="#GenomicBreeding.checkdims-Tuple{TEBV}"><code>GenomicBreeding.checkdims</code></a></li><li><a href="#GenomicBreeding.checkdims-Tuple{Genomes}"><code>GenomicBreeding.checkdims</code></a></li><li><a href="#GenomicBreeding.checkdims-Tuple{Phenomes}"><code>GenomicBreeding.checkdims</code></a></li><li><a href="#GenomicBreeding.countlevels-Tuple{DataFrames.DataFrame}"><code>GenomicBreeding.countlevels</code></a></li><li><a href="#GenomicBreeding.dimensions-Tuple{Genomes}"><code>GenomicBreeding.dimensions</code></a></li><li><a href="#GenomicBreeding.loci-Tuple{Genomes}"><code>GenomicBreeding.loci</code></a></li><li><a href="#GenomicBreeding.loci_alleles-Tuple{Genomes}"><code>GenomicBreeding.loci_alleles</code></a></li><li><a href="#GenomicBreeding.plot"><code>GenomicBreeding.plot</code></a></li><li><a href="#GenomicBreeding.readJLD2-Union{Tuple{Type{T}}, Tuple{T}, Tuple{Type{T}, String}} where T&lt;:GenomicBreeding.AbstractGB"><code>GenomicBreeding.readJLD2</code></a></li><li><a href="#GenomicBreeding.readdelimited-Tuple{Type{Genomes}}"><code>GenomicBreeding.readdelimited</code></a></li><li><a href="#GenomicBreeding.simulateeffects-Tuple{}"><code>GenomicBreeding.simulateeffects</code></a></li><li><a href="#GenomicBreeding.simulategenomes-Tuple{}"><code>GenomicBreeding.simulategenomes</code></a></li><li><a href="#GenomicBreeding.simulategenomiceffects-Tuple{}"><code>GenomicBreeding.simulategenomiceffects</code></a></li><li><a href="#GenomicBreeding.simulatetrials-Tuple{}"><code>GenomicBreeding.simulatetrials</code></a></li><li><a href="#GenomicBreeding.slice-Tuple{Genomes}"><code>GenomicBreeding.slice</code></a></li><li><a href="#GenomicBreeding.trialsmodelsfomulae!-Tuple{DataFrames.DataFrame}"><code>GenomicBreeding.trialsmodelsfomulae!</code></a></li><li><a href="#GenomicBreeding.writeJLD2-Tuple{GenomicBreeding.AbstractGB}"><code>GenomicBreeding.writeJLD2</code></a></li><li><a href="#GenomicBreeding.writedelimited-Tuple{Phenomes}"><code>GenomicBreeding.writedelimited</code></a></li><li><a href="#GenomicBreeding.writedelimited-Tuple{Trials}"><code>GenomicBreeding.writedelimited</code></a></li><li><a href="#GenomicBreeding.writedelimited-Tuple{Genomes}"><code>GenomicBreeding.writedelimited</code></a></li><li><a href="#GenomicBreeding.@string2formula-Tuple{Any}"><code>GenomicBreeding.@string2formula</code></a></li></ul><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicBreeding.Genomes" href="#GenomicBreeding.Genomes"><code>GenomicBreeding.Genomes</code></a> — <span class="docstring-category">Type</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><p><strong>Genomes struct</strong></p><p>Containes unique entries and loci_alleles where allele frequencies can have missing values</p><p><strong>Constructor</strong></p><pre><code class="language-julia hljs">Genomes(; n::Int64 = 1, p::Int64 = 2)</code></pre><p><strong>Fields</strong></p><ul><li><code>entries</code>: names of the <code>n</code> entries or samples</li><li><code>populations</code>: name/s of the population/s each entries or samples belong to</li><li><code>loci_alleles</code>: names of the <code>p</code> loci-alleles combinations (<code>p</code> = <code>l</code> loci x <code>a-1</code> alleles) including the </li></ul><p>chromsome or scaffold name, position, all alleles, and current allele separated by tabs (&quot;\t&quot;)</p><ul><li><code>allele_frequencies</code>: <code>n x p</code> matrix of allele frequencies between 0 and 1 which can have missing values</li><li><code>mask</code>: <code>n x p</code> matrix of boolean mask for selective analyses and slicing</li></ul><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; genomes = Genomes(n=2, p=2)
Genomes([&quot;&quot;, &quot;&quot;], [&quot;&quot;, &quot;&quot;], [&quot;&quot;, &quot;&quot;], Union{Missing, Float64}[missing missing; missing missing], Bool[0 0; 0 0])

julia&gt; fieldnames(Genomes)
(:entries, :populations, :loci_alleles, :allele_frequencies, :mask)

julia&gt; genomes.entries = [&quot;entry_1&quot;, &quot;entry_2&quot;];

julia&gt; genomes.populations = [&quot;pop_1&quot;, &quot;pop_1&quot;];

julia&gt; genomes.loci_alleles = [&quot;chr1\t12345\tA|T\tA&quot;, &quot;chr2\t678910\tC|D\tD&quot;];

julia&gt; genomes.allele_frequencies = [0.5 0.25; 0.9 missing];

julia&gt; genomes.mask = [true true; true false];

julia&gt; genomes
Genomes([&quot;entry_1&quot;, &quot;entry_2&quot;], [&quot;pop_1&quot;, &quot;pop_1&quot;], [&quot;chr1\t12345\tA|T\tA&quot;, &quot;chr2\t678910\tC|D\tD&quot;], Union{Missing, Float64}[0.5 0.25; 0.9 missing], Bool[1 1; 1 0])</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jeffersonfparil/GenomicBreeding.jl/blob/95e45ff7b84ffd98d4238d1241678add8e646bbd/src/core/genomes.jl#L1-L41">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicBreeding.Phenomes" href="#GenomicBreeding.Phenomes"><code>GenomicBreeding.Phenomes</code></a> — <span class="docstring-category">Type</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><p><strong>Phenomes struct</strong></p><p>Constains unique entries and traits where phenotype data can have missing values</p><p><strong>Constructor</strong></p><pre><code class="language-julia hljs">Phenomes(; n::Int64 = 1, t::Int64 = 2)</code></pre><p><strong>Fields</strong></p><ul><li><code>entries</code>: names of the <code>n</code> entries or samples</li><li><code>populations</code>: name/s of the population/s each entries or samples belong to</li><li><code>traits</code>: names of the <code>t</code> traits</li><li><code>phenotypes</code>: <code>n x t</code> matrix of numeric (<code>R</code>) phenotype data which can have missing values</li><li><code>mask</code>: <code>n x t</code> matrix of boolean mask for selective analyses and slicing</li></ul><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; phenomes = Phenomes(n=2, t=2)
Phenomes([&quot;&quot;, &quot;&quot;], [&quot;&quot;, &quot;&quot;], [&quot;&quot;, &quot;&quot;], Union{Missing, Float64}[missing missing; missing missing], Bool[0 0; 0 0])

julia&gt; fieldnames(Phenomes)
(:entries, :populations, :traits, :phenotypes, :mask)

julia&gt; phenomes.entries = [&quot;entry_1&quot;, &quot;entry_2&quot;];

julia&gt; phenomes.populations = [&quot;pop_A&quot;, &quot;pop_B&quot;];

julia&gt; phenomes.traits = [&quot;height&quot;, &quot;yield&quot;];

julia&gt; phenomes.phenotypes = [200.0 2.5; 150.0 missing];

julia&gt; phenomes.mask = [true true; true false];

julia&gt; phenomes
Phenomes([&quot;entry_1&quot;, &quot;entry_2&quot;], [&quot;pop_A&quot;, &quot;pop_B&quot;], [&quot;height&quot;, &quot;yield&quot;], Union{Missing, Float64}[200.0 2.5; 150.0 missing], Bool[1 1; 1 0])</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jeffersonfparil/GenomicBreeding.jl/blob/95e45ff7b84ffd98d4238d1241678add8e646bbd/src/core/phenomes.jl#L1-L40">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicBreeding.SimulatedEffects" href="#GenomicBreeding.SimulatedEffects"><code>GenomicBreeding.SimulatedEffects</code></a> — <span class="docstring-category">Type</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><p><strong>SimulatedEffects struct</strong></p><p>Contains:</p><ol><li>Identification, i.e. trait, year, season, harvest, site, and replication</li><li>Additive environmental effects:<ul><li>year</li><li>season</li><li>site</li></ul></li><li>Environmental interaction effects:<ul><li>seasons<em>x</em>year</li><li>harvests<em>x</em>season<em>x</em>year</li><li>sites<em>x</em>harvest<em>x</em>season<em>x</em>year</li></ul></li><li>Spatial effects including the field layout per year-season-harvest-site combination<ul><li>field_layout</li><li>replications<em>x</em>site<em>x</em>harvest<em>x</em>season<em>x</em>year</li><li>blocks<em>x</em>site<em>x</em>harvest<em>x</em>season<em>x</em>year</li><li>rows<em>x</em>site<em>x</em>harvest<em>x</em>season<em>x</em>year</li><li>cols<em>x</em>site<em>x</em>harvest<em>x</em>season<em>x</em>year</li></ul></li><li>Genetic effects<ul><li>additive_genetic</li><li>dominance_genetic</li><li>epistasis_genetic</li></ul></li><li>GxE effects<ul><li>additive<em>allele</em>x<em>site</em>x<em>harvest</em>x<em>season</em>x_year</li><li>dominance<em>allele</em>x<em>site</em>x<em>harvest</em>x<em>season</em>x_year</li><li>epistasis<em>allele</em>x<em>site</em>x<em>harvest</em>x<em>season</em>x_year</li></ul></li></ol><p><strong>Constructor</strong></p><pre><code class="language-julia hljs">SimulatedEffects()</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jeffersonfparil/GenomicBreeding.jl/blob/95e45ff7b84ffd98d4238d1241678add8e646bbd/src/simulation/simulate_effects.jl#L1-L34">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicBreeding.TEBV" href="#GenomicBreeding.TEBV"><code>GenomicBreeding.TEBV</code></a> — <span class="docstring-category">Type</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><p><strong>Trial-estimate breeding values (TEBV) struct</strong></p><p>Contains trial-estimated breeding values as generated by <code>analyse(trials::Trials, ...)</code>.</p><p><strong>Fields</strong></p><ul><li><code>traits</code>: names of the traits <code>t</code> traits</li><li><code>formulae</code>: best-fitting formula for each trait</li><li><code>models</code>: best-fitting linear mixed model for each trait</li><li><code>df_BLUEs</code>: best linear unbiased estimators or fixed effects table of each best fitting model</li><li><code>df_BLUPs</code>: best linear unbiased predictors or random effects table of each best fitting model</li><li><code>df_TEBVs</code>: breeding values of each entry per trait</li></ul><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; tebv = TEBV(traits=[&quot;&quot;], formulae=[&quot;&quot;], models=[MixedModel(@formula(y~1+(1|x)), DataFrame(y=1, x=1))], df_BLUEs=[DataFrame(x=1)], df_BLUPs=[DataFrame(x=1)], df_TEBVs=[DataFrame(x=1)]);

julia&gt; tebv.traits
1-element Vector{String}:
 &quot;&quot;</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jeffersonfparil/GenomicBreeding.jl/blob/95e45ff7b84ffd98d4238d1241678add8e646bbd/src/models/tebv.jl#L1-L22">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicBreeding.Trials" href="#GenomicBreeding.Trials"><code>GenomicBreeding.Trials</code></a> — <span class="docstring-category">Type</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><p><strong>Trials struct</strong></p><p>Contains phenotype data across years, seasons, harvest, sites, populations, replications, blocks, rows, and columns</p><p><strong>Constructor</strong></p><pre><code class="language-julia hljs">Trials(; n::Int64 = 2, p::Int64 = 2)</code></pre><p><strong>Fields</strong></p><ul><li><code>phenotypes</code>: <code>n x t</code> matrix of numeric phenotype data which can have missing values</li><li><code>traits</code>: names of the traits <code>t</code> traits</li><li><code>years</code>: names of the years corresponding to each row in the phenotype matrix</li><li><code>seasons</code>: names of the seasons corresponding to each row in the phenotype matrix</li><li><code>harvests</code>: names of the harvests corresponding to each row in the phenotype matrix</li><li><code>sites</code>: names of the sites corresponding to each row in the phenotype matrix</li><li><code>replications</code>: names of the replications corresponding to each row in the phenotype matrix</li><li><code>blocks</code>: names of the blocks corresponding to each row in the phenotype matrix</li><li><code>rows</code>: names of the rows corresponding to each row in the phenotype matrix</li><li><code>cols</code>: names of the cols corresponding to each row in the phenotype matrix</li><li><code>entries</code>: names of the entries corresponding to each row in the phenotype matrix</li><li><code>populations</code>: names of the populations corresponding to each row in the phenotype matrix</li></ul><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; trials = Trials(n=1, t=2)
Trials(Union{Missing, Float64}[missing missing], [&quot;&quot;, &quot;&quot;], [&quot;&quot;], [&quot;&quot;], [&quot;&quot;], [&quot;&quot;], [&quot;&quot;], [&quot;&quot;], [&quot;&quot;], [&quot;&quot;], [&quot;&quot;], [&quot;&quot;])

julia&gt; fieldnames(Trials)
(:phenotypes, :traits, :years, :seasons, :harvests, :sites, :replications, :blocks, :rows, :cols, :entries, :populations)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jeffersonfparil/GenomicBreeding.jl/blob/95e45ff7b84ffd98d4238d1241678add8e646bbd/src/core/trials.jl#L1-L34">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="Base.:==-Tuple{Genomes, Genomes}" href="#Base.:==-Tuple{Genomes, Genomes}"><code>Base.:==</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">Base.:(==)(x::Genomes, y::Genomes)::Bool</code></pre><p>Equality of Genomes structs using the hash function defined for Genomes structs.</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; genomes_1 = genomes = Genomes(n=2,p=4);

julia&gt; genomes_2 = genomes = Genomes(n=2,p=4);

julia&gt; genomes_3 = genomes = Genomes(n=1,p=2);

julia&gt; genomes_1 == genomes_2
true

julia&gt; genomes_1 == genomes_3
false</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jeffersonfparil/GenomicBreeding.jl/blob/95e45ff7b84ffd98d4238d1241678add8e646bbd/src/core/genomes.jl#L73-L92">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="Base.:==-Tuple{Phenomes, Phenomes}" href="#Base.:==-Tuple{Phenomes, Phenomes}"><code>Base.:==</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">Base.:(==)(x::Phenomes, y::Phenomes)::Bool</code></pre><p>Equality of Phenomes structs using the hash function defined for Phenomes structs.</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; phenomes_1 = phenomes = Phenomes(n=2, t=4);

julia&gt; phenomes_2 = phenomes = Phenomes(n=2, t=4);

julia&gt; phenomes_3 = phenomes = Phenomes(n=1, t=2);

julia&gt; phenomes_1 == phenomes_2
true

julia&gt; phenomes_1 == phenomes_3
false</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jeffersonfparil/GenomicBreeding.jl/blob/95e45ff7b84ffd98d4238d1241678add8e646bbd/src/core/phenomes.jl#L71-L90">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="Base.:==-Tuple{SimulatedEffects, SimulatedEffects}" href="#Base.:==-Tuple{SimulatedEffects, SimulatedEffects}"><code>Base.:==</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">Base.:(==)(x::SimulatedEffects, y::SimulatedEffects)::Bool</code></pre><p>Equality of SimulatedEffects structs using the hash function defined for SimulatedEffects structs.</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; effects_1 = SimulatedEffects();

julia&gt; effects_2 = SimulatedEffects();

julia&gt; effects_3 = SimulatedEffects(); effects_3.id[1] = &quot;SOMETHING_ELSE&quot;;

julia&gt; effects_1 == effects_2
true

julia&gt; effects_1 == effects_3
false</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jeffersonfparil/GenomicBreeding.jl/blob/95e45ff7b84ffd98d4238d1241678add8e646bbd/src/simulation/simulate_effects.jl#L154-L173">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="Base.:==-Tuple{TEBV, TEBV}" href="#Base.:==-Tuple{TEBV, TEBV}"><code>Base.:==</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">Base.:(==)(x::TEBV, y::TEBV)::Bool</code></pre><p>Equality of TEBV structs using the hash function defined for TEBV structs.</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; tebv_1 = TEBV(traits=[&quot;&quot;], formulae=[&quot;&quot;], models=[MixedModel(@formula(y~1+(1|x)), DataFrame(y=1, x=1))], df_BLUEs=[DataFrame(x=1)], df_BLUPs=[DataFrame(x=1)], df_TEBVs=[DataFrame(x=1)]);

julia&gt; tebv_2 = TEBV(traits=[&quot;&quot;], formulae=[&quot;&quot;], models=[MixedModel(@formula(y~1+(1|x)), DataFrame(y=1, x=1))], df_BLUEs=[DataFrame(x=1)], df_BLUPs=[DataFrame(x=1)], df_TEBVs=[DataFrame(x=1)]);

julia&gt; tebv_3 = TEBV(traits=[&quot;SOMETHING_ELSE&quot;], formulae=[&quot;&quot;], models=[MixedModel(@formula(y~1+(1|x)), DataFrame(y=1, x=1))], df_BLUEs=[DataFrame(x=1)], df_BLUPs=[DataFrame(x=1)], df_TEBVs=[DataFrame(x=1)]);

julia&gt; tebv_1 == tebv_2
true

julia&gt; tebv_1 == tebv_3
false</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jeffersonfparil/GenomicBreeding.jl/blob/95e45ff7b84ffd98d4238d1241678add8e646bbd/src/models/tebv.jl#L63-L82">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="Base.:==-Tuple{Trials, Trials}" href="#Base.:==-Tuple{Trials, Trials}"><code>Base.:==</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">Base.:(==)(x::Trials, y::Trials)::Bool</code></pre><p>Equality of Trials structs using the hash function defined for Trials structs.</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; trials_1 = trials = Trials(n=2, t=4);

julia&gt; trials_2 = trials = Trials(n=2, t=4);

julia&gt; trials_3 = trials = Trials(n=1, t=2);

julia&gt; trials_1 == trials_2
true

julia&gt; trials_1 == trials_3
false</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jeffersonfparil/GenomicBreeding.jl/blob/95e45ff7b84ffd98d4238d1241678add8e646bbd/src/core/trials.jl#L109-L128">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="Base.hash-Tuple{Genomes, UInt64}" href="#Base.hash-Tuple{Genomes, UInt64}"><code>Base.hash</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">Base.hash(x::Genomes, h::UInt)::UInt</code></pre><p>Hash a Genomes struct using the entries, populations and loci<em>alleles. We deliberately excluded the allele</em>frequencies, and mask for efficiency.</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; genomes = Genomes(n=2, p=2);

julia&gt; typeof(hash(genomes))
UInt64</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jeffersonfparil/GenomicBreeding.jl/blob/95e45ff7b84ffd98d4238d1241678add8e646bbd/src/core/genomes.jl#L53-L66">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="Base.hash-Tuple{Phenomes, UInt64}" href="#Base.hash-Tuple{Phenomes, UInt64}"><code>Base.hash</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">Base.hash(x::Phenomes, h::UInt)::UInt</code></pre><p>Hash a Phenomes struct.</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; phenomes = Phenomes(n=2, t=2);

julia&gt; typeof(hash(phenomes))
UInt64</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jeffersonfparil/GenomicBreeding.jl/blob/95e45ff7b84ffd98d4238d1241678add8e646bbd/src/core/phenomes.jl#L53-L65">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="Base.hash-Tuple{SimulatedEffects, UInt64}" href="#Base.hash-Tuple{SimulatedEffects, UInt64}"><code>Base.hash</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">Base.hash(x::SimulatedEffects, h::UInt)::UInt</code></pre><p>Hash a SimulatedEffects struct.</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; effects = SimulatedEffects();

julia&gt; typeof(hash(effects))
UInt64</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jeffersonfparil/GenomicBreeding.jl/blob/95e45ff7b84ffd98d4238d1241678add8e646bbd/src/simulation/simulate_effects.jl#L79-L91">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="Base.hash-Tuple{TEBV, UInt64}" href="#Base.hash-Tuple{TEBV, UInt64}"><code>Base.hash</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">Base.hash(x::TEBV, h::UInt)::UInt</code></pre><p>Hash a TEBV struct using the traits, formualae and df<em>TEBVs. We deliberately excluded the models, df</em>BLUEs, and df_BLUPs for efficiency.</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; tebv = TEBV(traits=[&quot;&quot;], formulae=[&quot;&quot;], models=[MixedModel(@formula(y~1+(1|x)), DataFrame(y=1, x=1))], df_BLUEs=[DataFrame(x=1)], df_BLUPs=[DataFrame(x=1)], df_TEBVs=[DataFrame(x=1)]);

julia&gt; typeof(hash(tebv))
UInt64</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jeffersonfparil/GenomicBreeding.jl/blob/95e45ff7b84ffd98d4238d1241678add8e646bbd/src/models/tebv.jl#L43-L56">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="Base.hash-Tuple{Trials, UInt64}" href="#Base.hash-Tuple{Trials, UInt64}"><code>Base.hash</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">Base.hash(x::Trials, h::UInt)::UInt</code></pre><p>Hash a Trials struct.</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; trials = Trials(n=2, t=2);

julia&gt; typeof(hash(trials))
UInt64</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jeffersonfparil/GenomicBreeding.jl/blob/95e45ff7b84ffd98d4238d1241678add8e646bbd/src/core/trials.jl#L67-L79">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="Base.sum-Tuple{SimulatedEffects}" href="#Base.sum-Tuple{SimulatedEffects}"><code>Base.sum</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">sum(effects::SimulatedEffects)::Tuple{Int64, Int64, Int64}</code></pre><p>Sum up the simulated effects to generate the simulated phenotype values</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; effects = SimulatedEffects();

julia&gt; sum(effects)
1-element Vector{Float64}:
 0.0

julia&gt; effects.additive_genetic[1] = pi;

julia&gt; sum(effects)
1-element Vector{Float64}:
 3.141592653589793</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jeffersonfparil/GenomicBreeding.jl/blob/95e45ff7b84ffd98d4238d1241678add8e646bbd/src/simulation/simulate_effects.jl#L215-L234">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicBreeding.analyse-Tuple{Trials}" href="#GenomicBreeding.analyse-Tuple{Trials}"><code>GenomicBreeding.analyse</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><p><strong>Analyse trials</strong></p><p><strong>Arguments</strong></p><ul><li><code>trials</code>: Trials struct </li><li><code>max_levels</code>: maximum number of non-entry factor levels to include in the linear mixed models (default = 100)</li><li><code>max_time_per_model</code>: maximum time in seconds for fitting each linear mixed model (default = 60)</li><li><code>verbose</code>: Show trials analysis progress bar? (default = true)</li></ul><p><strong>Outputs</strong></p><ul><li><code>TEBV</code> struct containing the trait names, the best fitting formulae, models, BLUEs, and BLUPs for each trait</li></ul><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; trials, _simulated_effects = simulatetrials(genomes = simulategenomes(n=10, verbose=false), n_years=1, n_seasons=1, n_harvests=1, n_sites=1, n_replications=10, verbose=false);

julia&gt; tebv = analyse(trials, max_levels=50, verbose=false);

julia&gt; tebv.traits
3-element Vector{String}:
 &quot;trait_1&quot;
 &quot;trait_2&quot;
 &quot;trait_3&quot;

julia&gt; checkdims(tebv)
true</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jeffersonfparil/GenomicBreeding.jl/blob/95e45ff7b84ffd98d4238d1241678add8e646bbd/src/models/tebv.jl#L380-L407">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicBreeding.checkdims-Tuple{Genomes}" href="#GenomicBreeding.checkdims-Tuple{Genomes}"><code>GenomicBreeding.checkdims</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">checkdims(genomes::Genomes)::Bool</code></pre><p>Check dimension compatibility of the fields of the Genomes struct</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; genomes = Genomes(n=2,p=4);

julia&gt; checkdims(genomes)
false

julia&gt; genomes.entries = [&quot;entry_1&quot;, &quot;entry_2&quot;];

julia&gt; genomes.loci_alleles = [&quot;chr1\t1\tA|T\tA&quot;, &quot;chr1\t2\tC|G\tG&quot;, &quot;chr2\t3\tA|T\tA&quot;, &quot;chr2\t4\tG|T\tG&quot;];

julia&gt; checkdims(genomes)
true</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jeffersonfparil/GenomicBreeding.jl/blob/95e45ff7b84ffd98d4238d1241678add8e646bbd/src/core/genomes.jl#L98-L117">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicBreeding.checkdims-Tuple{Phenomes}" href="#GenomicBreeding.checkdims-Tuple{Phenomes}"><code>GenomicBreeding.checkdims</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">checkdims(y::Phenomes)::Bool</code></pre><p>Check dimension compatibility of the fields of the Phenomes struct</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; y = Phenomes(n=2, t=2);

julia&gt; checkdims(y)
false

julia&gt; y.entries = [&quot;entry_1&quot;, &quot;entry_2&quot;];

julia&gt; y.traits = [&quot;trait_1&quot;, &quot;trait_2&quot;];

julia&gt; checkdims(y)
true</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jeffersonfparil/GenomicBreeding.jl/blob/95e45ff7b84ffd98d4238d1241678add8e646bbd/src/core/phenomes.jl#L96-L115">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicBreeding.checkdims-Tuple{SimulatedEffects}" href="#GenomicBreeding.checkdims-Tuple{SimulatedEffects}"><code>GenomicBreeding.checkdims</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">checkdims(effects::SimulatedEffects)::Bool</code></pre><p>Check dimension compatibility of the fields of the SimulatedEffects struct</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; effects = SimulatedEffects();

julia&gt; checkdims(effects)
true

julia&gt; effects.id = [&quot;beaking_change&quot;];

julia&gt; checkdims(effects)
false</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jeffersonfparil/GenomicBreeding.jl/blob/95e45ff7b84ffd98d4238d1241678add8e646bbd/src/simulation/simulate_effects.jl#L179-L196">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicBreeding.checkdims-Tuple{TEBV}" href="#GenomicBreeding.checkdims-Tuple{TEBV}"><code>GenomicBreeding.checkdims</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">checkdims(y::TEBV)::Bool</code></pre><p>Check dimension compatibility of the fields of the TEBV struct</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; tebv = TEBV(traits=[&quot;&quot;], formulae=[&quot;&quot;], models=[MixedModel(@formula(y~1+(1|x)), DataFrame(y=1, x=1))], df_BLUEs=[DataFrame(x=1)], df_BLUPs=[DataFrame(x=1)], df_TEBVs=[DataFrame(x=1)]);

julia&gt; checkdims(tebv)
true</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jeffersonfparil/GenomicBreeding.jl/blob/95e45ff7b84ffd98d4238d1241678add8e646bbd/src/models/tebv.jl#L88-L100">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicBreeding.checkdims-Tuple{Trials}" href="#GenomicBreeding.checkdims-Tuple{Trials}"><code>GenomicBreeding.checkdims</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">checkdims(trials::Trials)::Bool</code></pre><p>Check dimension compatibility of the fields of the Trials struct</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; trials = Trials(n=1, t=2);

julia&gt; trials.entries = [&quot;entry_1&quot;];

julia&gt; checkdims(trials)
true

julia&gt; trials.entries = [&quot;entering_2_entries&quot;, &quot;instead_of_just_1&quot;];

julia&gt; checkdims(trials)
false</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jeffersonfparil/GenomicBreeding.jl/blob/95e45ff7b84ffd98d4238d1241678add8e646bbd/src/core/trials.jl#L134-L153">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicBreeding.countlevels-Tuple{DataFrames.DataFrame}" href="#GenomicBreeding.countlevels-Tuple{DataFrames.DataFrame}"><code>GenomicBreeding.countlevels</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">countlevels(df::DataFrame; column_names::Vector{String})::Int64</code></pre><p>Count the total number of factor levels across the column names provided.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jeffersonfparil/GenomicBreeding.jl/blob/95e45ff7b84ffd98d4238d1241678add8e646bbd/src/models/tebv.jl#L122-L126">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicBreeding.dimensions-Tuple{Genomes}" href="#GenomicBreeding.dimensions-Tuple{Genomes}"><code>GenomicBreeding.dimensions</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">dimensions(genomes::Genomes)::Dict{String, Int64}</code></pre><p>Count the number of entries, populations, loci-alleles combination, loci, and maximum number of alleles per locus in the Genomes struct</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; genomes = simulategenomes(n=100, l=1_000, n_alleles=4, verbose=false);

julia&gt; dimensions(genomes)
Dict{String, Int64} with 6 entries:
  &quot;n_entries&quot;      =&gt; 100
  &quot;n_chr&quot;          =&gt; 7
  &quot;n_loci&quot;         =&gt; 1000
  &quot;n_loci_alleles&quot; =&gt; 3000
  &quot;n_populations&quot;  =&gt; 1
  &quot;max_n_alleles&quot;  =&gt; 4</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jeffersonfparil/GenomicBreeding.jl/blob/95e45ff7b84ffd98d4238d1241678add8e646bbd/src/core/genomes.jl#L138-L156">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicBreeding.loci-Tuple{Genomes}" href="#GenomicBreeding.loci-Tuple{Genomes}"><code>GenomicBreeding.loci</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">loci(genomes::Genomes)::Tuple{Vector{String},Vector{Int64},Vector{Int64},Vector{Int64}}</code></pre><p>Extract chromosome names, positions, start and end indexes of each locus across loci</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; genomes = simulategenomes(n=100, l=1_000, n_alleles=4, verbose=false);

julia&gt; chromsomes, positions, loci_ini_idx, loci_fin_idx = loci(genomes);

julia&gt; length(chromsomes), length(positions), length(loci_ini_idx), length(loci_fin_idx)
(1000, 1000, 1000, 1000)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jeffersonfparil/GenomicBreeding.jl/blob/95e45ff7b84ffd98d4238d1241678add8e646bbd/src/core/genomes.jl#L227-L241">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicBreeding.loci_alleles-Tuple{Genomes}" href="#GenomicBreeding.loci_alleles-Tuple{Genomes}"><code>GenomicBreeding.loci_alleles</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">loci_alleles(genomes::Genomes)::Tuple{Vector{String},Vector{Int64},Vector{String}}</code></pre><p>Extract chromosomes, positions, and alleles across loci-allele combinations</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; genomes = simulategenomes(n=100, l=1_000, n_alleles=4, verbose=false);

julia&gt; chromsomes, positions, alleles = loci_alleles(genomes);

julia&gt; length(chromsomes), length(positions), length(alleles)
(3000, 3000, 3000)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jeffersonfparil/GenomicBreeding.jl/blob/95e45ff7b84ffd98d4238d1241678add8e646bbd/src/core/genomes.jl#L198-L212">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicBreeding.plot" href="#GenomicBreeding.plot"><code>GenomicBreeding.plot</code></a> — <span class="docstring-category">Function</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">plot(genomes::Genomes)::Nothing</code></pre><p>Plot allele frequencies</p><p><strong>Examples</strong></p><pre><code class="nohighlight hljs">julia&gt; genomes = simulategenomes(n=100, l=1_000, n_alleles=4, verbose=false);

julia&gt; GenomicBreeding.plot(genomes);
</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jeffersonfparil/GenomicBreeding.jl/blob/95e45ff7b84ffd98d4238d1241678add8e646bbd/src/core/genomes.jl#L272-L284">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicBreeding.readJLD2-Union{Tuple{Type{T}}, Tuple{T}, Tuple{Type{T}, String}} where T&lt;:GenomicBreeding.AbstractGB" href="#GenomicBreeding.readJLD2-Union{Tuple{Type{T}}, Tuple{T}, Tuple{Type{T}, String}} where T&lt;:GenomicBreeding.AbstractGB"><code>GenomicBreeding.readJLD2</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">readJLD2(type::Type, fname::String = missing)::Type</code></pre><p>Load a core (<code>Genomes</code>, <code>Phenomes</code>, and <code>Trials</code>), simulation (<code>SimulatedEffects</code>), and model (<code>TEBV</code>) struct from a JLD2 file.</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; genomes = simulategenomes(n=2, verbose=false);

julia&gt; fname = writeJLD2(genomes);

julia&gt; readJLD2(Genomes, fname) == genomes
true

julia&gt; phenomes = Phenomes(n=2, t=2); phenomes.entries = [&quot;entry_1&quot;, &quot;entry_2&quot;]; phenomes.traits = [&quot;trait_1&quot;, &quot;trait_2&quot;];

julia&gt; fname = writeJLD2(phenomes);

julia&gt; readJLD2(Phenomes, fname) == phenomes
true

julia&gt; trials = Trials(n=1, t=2); trials.entries = [&quot;entry_1&quot;];

julia&gt; fname = writeJLD2(trials);

julia&gt; readJLD2(Trials, fname) == trials
true</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jeffersonfparil/GenomicBreeding.jl/blob/95e45ff7b84ffd98d4238d1241678add8e646bbd/src/io/reader.jl#L1-L29">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicBreeding.readdelimited-Tuple{Type{Genomes}}" href="#GenomicBreeding.readdelimited-Tuple{Type{Genomes}}"><code>GenomicBreeding.readdelimited</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">readdelimited(type::Type{Genomes}; fname::String, sep::String = &quot;\t&quot;)::Genomes</code></pre><p>Load a <code>Genomes</code> struct from a string-delimited (default=<code>\t</code>) file.  Each row corresponds to a locus-allele combination. The first 4 columns correspond to the chromosome, position, all alleles in the locus (delimited by <code>|</code>), and the specific allele. The subsequency columns refer to the samples, pools, entries or genotypes.</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; genomes = simulategenomes(n=10, verbose=false);

julia&gt; fname = writedelimited(genomes);

julia&gt; genomes_reloaded = readdelimited(Genomes, fname=fname);

julia&gt; genomes == genomes_reloaded
true</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jeffersonfparil/GenomicBreeding.jl/blob/95e45ff7b84ffd98d4238d1241678add8e646bbd/src/io/reader.jl#L44-L63">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicBreeding.simulateeffects-Tuple{}" href="#GenomicBreeding.simulateeffects-Tuple{}"><code>GenomicBreeding.simulateeffects</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><p><strong>Simulate effects</strong></p><p>Sample <code>p</code> x <code>q</code> effects from a multivariate normal distribution with <code>μ~Exp(λ)</code> and <code>Σ=μμ&#39;</code></p><p><strong>Arguments</strong></p><ul><li><code>p</code>: number of correlated effects to simulate (default = 2)</li><li><code>q</code>: number times to simulate the correlated effects from the same distribution (default = 1)</li><li><code>λ</code>: parameter of the exponential distritbution from which the means will be sampled from (default = 1.00)</li><li><code>seed</code>: randomisation seed (default = 42)</li></ul><p><strong>Output</strong></p><ul><li><code>p</code> x <code>q</code> matrix of correlated effects</li></ul><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; θ::Matrix{Float64} = simulateeffects();

julia&gt; sum(abs.(θ - [-0.0886501800782904; -0.596478483888422])) &lt; 0.00001
true</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jeffersonfparil/GenomicBreeding.jl/blob/95e45ff7b84ffd98d4238d1241678add8e646bbd/src/simulation/simulate_effects.jl#L246-L268">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicBreeding.simulategenomes-Tuple{}" href="#GenomicBreeding.simulategenomes-Tuple{}"><code>GenomicBreeding.simulategenomes</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><p><strong>Simulate genomes</strong></p><p><strong>Arguments</strong></p><ul><li><code>n</code>: number of entries (default = 100)</li><li><code>l</code>: number of loci (default = 10_000)</li><li><code>n_chroms</code>: number of chromosomes (default = 7)</li><li><code>n_alleles</code>: number of alleles per locus (default = 2)</li><li><code>max_pos</code>: total length of the genome in base-pairs (bp) (default = 135<em>000</em>000)</li><li><code>ld_corr_50perc_kb</code>: distance in bp at which linkage expressed as correlation between a pair of loci is at 50% (default = 1_000)</li><li><code>μ_β_params</code>: the shape parameters of the Beta distribution from which the mean allele frequencies will be sampled  (default = (0.5, 0.5); U-shaped distribution; you may use (2.0, 2.0) for a bell-shaped distribution)</li><li><code>sparsity</code>: Proportion of missing data (default = 0.0)</li><li><code>seed</code>: psuedo-random number generator seed for replicability (default = 42)</li><li><code>verbose</code>: Show progress bar and plot the linkage heatmap into an svg file? (default = true)</li></ul><p><strong>Output</strong></p><ul><li><code>Genomes</code></li></ul><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; genomes = simulategenomes(n=100, l=10_000, n_alleles=3, verbose=false);

julia&gt; length(genomes.entries)
100

julia&gt; length(genomes.populations)
100

julia&gt; length(genomes.loci_alleles)
20000

julia&gt; size(genomes.allele_frequencies)
(100, 20000)

julia&gt; mean(ismissing.(genomes.allele_frequencies))
0.0

julia&gt; rng::TaskLocalRNG = Random.seed!(123);

julia&gt; idx = StatsBase.sample(rng, range(1, 20_000, step=2), 250, replace = false, ordered = true);

julia&gt; correlations = StatsBase.cor(genomes.allele_frequencies[:, idx]);

julia&gt; correlations[10,10] == 1.00
true

julia&gt; correlations[10,10] &gt; correlations[10,250]
true

julia&gt; genomes = simulategenomes(n=100, l=10_000, n_alleles=3, sparsity=0.25, verbose=false);

julia&gt; mean(ismissing.(genomes.allele_frequencies))
0.25</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jeffersonfparil/GenomicBreeding.jl/blob/95e45ff7b84ffd98d4238d1241678add8e646bbd/src/simulation/simulate_genomes.jl#L1-L56">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicBreeding.simulategenomiceffects-Tuple{}" href="#GenomicBreeding.simulategenomiceffects-Tuple{}"><code>GenomicBreeding.simulategenomiceffects</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><p><strong>Simulate genomic effects</strong></p><p>Simulate additive, dominance, and epistatic effects</p><p><strong>Arguments</strong></p><ul><li><code>genomes</code>: Genome struct includes the <code>n</code> entries x <code>p</code> loci-alleles combinations (<code>p</code> = <code>l</code> loci x <code>a-1</code> alleles)</li><li><code>f_additive</code>: proportion of the <code>l</code> loci with non-zero additive effects on the phenotype</li><li><code>f_dominance</code>: proportion of the <code>l*f_additive</code> additive effects loci with additional dominance effects</li><li><code>f_epistasis</code>: proportion of the <code>l*f_additive</code> additive effects loci with additional epistasis effects</li></ul><p><strong>Outputs</strong></p><ul><li><code>n</code> x <code>3</code> matrix of additive, dominance and epistasis effects per entry</li><li><code>p</code> x <code>3</code> matrix of additive, dominance and epistasis effects per locus-allele combination</li></ul><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; genomes::Genomes = simulategenomes(n=100, l=2_000, n_alleles=3, verbose=false);

julia&gt; G, B = simulategenomiceffects(genomes=genomes, f_additive=0.05, f_dominance=0.75, f_epistasis=0.25);

julia&gt; size.([G, B])
2-element Vector{Tuple{Int64, Int64}}:
 (100, 3)
 (4000, 3)

julia&gt; sum(B .!= 0.0, dims=1)
1×3 Matrix{Int64}:
 200  75  50</code></pre><p><strong>Details</strong></p><p>The additive, dominance, and epistasis allele effects share a common exponential distribution (<code>λ=1</code>) from which  the mean of the effects (<code>μ</code>) are sampled, and the covariance matrix is derived (<code>Σ = μ * μ&#39;</code>;  where if <code>det(Σ)≈0</code> then we iteratively add 1.00 to the diagonals until it becomes invertible or 10 iterations  finishes and throws an error). The non-additive or epistasis allele effects were simulated by multiplying the allele  frequencies of all possible unique pairs of epistasis alleles and their effects.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jeffersonfparil/GenomicBreeding.jl/blob/95e45ff7b84ffd98d4238d1241678add8e646bbd/src/simulation/simulate_effects.jl#L291-L328">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicBreeding.simulatetrials-Tuple{}" href="#GenomicBreeding.simulatetrials-Tuple{}"><code>GenomicBreeding.simulatetrials</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><p><strong>Simulate trials</strong></p><p><strong>Arguments</strong></p><ul><li><code>genomes</code>: Genome struct includes the <code>n</code> entries x <code>p</code> loci-alleles combinations (<code>p</code> = <code>l</code> loci x <code>a-1</code> alleles)</li><li><code>f_add_dom_epi</code>: <code>n_traits</code> x 3 numeric matrix of loci proportion with additive, dominance and epistasis effects, i.e. each column refers to:<ul><li><code>f_additive</code>: proportion of the <code>l</code> loci with non-zero additive effects on the phenotype</li><li><code>f_dominance</code>: proportion of the <code>l*f_additive</code> additive effects loci with additional dominance effects</li><li><code>f_epistasis</code>: proportion of the <code>l*f_additive</code> additive effects loci with additional epistasis effects</li></ul></li><li><code>n_years</code>: Number of years (default = 2)</li><li><code>n_seasons</code>: Number of seasons (default = 4)</li><li><code>n_harvests</code>: Number of harvests (default = 2)</li><li><code>n_sites</code>: Number of sites (default = 4)</li><li><code>n_replications</code>: Number of replications (default = 2)</li><li><code>n_blocks</code>: Number of blocks across the entire field layout (default = missing)</li><li><code>n_rows</code>: Number of rows across the entire field layout (default = missing)</li><li><code>n_cols</code>: Number of columns across the entire field layout (default = missing)</li><li><code>proportion_of_variance</code>: <code>9</code> x <code>n_traits</code> numeric matrix of scaled/non-scaled proportion of variances allocated to   genetic and environmental effects (default = missing; values will be sampled from a uniform distribution  followed by a biased sample on the first row, i.e. additive effects row).  The rows correspond to the variance allocated to:<ol><li>additive genetic effects</li><li>dominance genetic effects</li><li>epistasis genetic effects</li><li>years effects</li><li>seasons effects</li><li>sites effects</li><li>environmental interactions</li><li>spatial interactions</li><li>GxE interactiions</li></ol></li><li><code>seed</code>: Randomisation seed (default = 42)</li><li><code>sparsity</code>: Proportion of missing data (default = 0.0)</li><li><code>verbose</code>: Show trials simulation progress bar? (default = true)</li></ul><p><strong>Outputs</strong></p><ul><li><code>Trials</code> struct of simulated phenotype data</li><li>Vector of <code>SimulatedEffects</code> each corresponding to each trait-year-season-harvest-site-replication combination</li></ul><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; genomes::Genomes = simulategenomes(n=100, l=2_000, n_alleles=3, verbose=false);

julia&gt; trials::Trials, vector_of_effects::Array{SimulatedEffects,1} = simulatetrials(genomes=genomes, sparsity=0.25, verbose=false);

julia&gt; size(trials.phenotypes)
(12800, 3)

julia&gt; size(trials.traits)
(3,)

julia&gt; unique(trials.entries) == genomes.entries
true

julia&gt; unique(trials.populations) == unique(genomes.populations)
true

julia&gt; abs(mean(ismissing.(trials.phenotypes)) - 0.25) &lt; 0.00001
true</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jeffersonfparil/GenomicBreeding.jl/blob/95e45ff7b84ffd98d4238d1241678add8e646bbd/src/simulation/simulate_trials.jl#L1-L62">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicBreeding.slice-Tuple{Genomes}" href="#GenomicBreeding.slice-Tuple{Genomes}"><code>GenomicBreeding.slice</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">slice(genomes::Genomes;idx_entries::Vector{Int64},idx_loci_alleles::Vector{Int64})::Genomes</code></pre><p>Count the number of entries, populations, loci, and maximum number of alleles per locus in the Genomes struct</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; genomes = simulategenomes(n=100, l=1_000, n_alleles=4, verbose=false);

julia&gt; sliced_genomes = slice(genomes, idx_entries=collect(1:10); idx_loci_alleles=collect(1:300));

julia&gt; dimensions(sliced_genomes)
Dict{String, Int64} with 6 entries:
  &quot;n_entries&quot;      =&gt; 10
  &quot;n_chr&quot;          =&gt; 1
  &quot;n_loci&quot;         =&gt; 100
  &quot;n_loci_alleles&quot; =&gt; 300
  &quot;n_populations&quot;  =&gt; 1
  &quot;max_n_alleles&quot;  =&gt; 4</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jeffersonfparil/GenomicBreeding.jl/blob/95e45ff7b84ffd98d4238d1241678add8e646bbd/src/core/genomes.jl#L331-L351">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicBreeding.trialsmodelsfomulae!-Tuple{DataFrames.DataFrame}" href="#GenomicBreeding.trialsmodelsfomulae!-Tuple{DataFrames.DataFrame}"><code>GenomicBreeding.trialsmodelsfomulae!</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">trialsmodelsfomulae!(df::DataFrame; trait::String, max_levels::Int64 = 100)::Vector{String}</code></pre><p>Define formulae for the mixed models to fit on the tabularised <code>Trials</code> struct.     - appends interaction effects intto <code>df</code>     - returns:         + a vector of formulae as strings         + a vector of the total number of non-entry factor levels</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; trials, _simulated_effects = simulatetrials(genomes = simulategenomes(verbose=false), verbose=false);

julia&gt; df = tabularise(trials);

julia&gt; size(df)
(12800, 14)

julia&gt; formulae, n_levels = trialsmodelsfomulae!(df, trait=&quot;trait_1&quot;);

julia&gt; size(df)
(12800, 134)

julia&gt; length(formulae)
76

julia&gt; sum(n_levels .== sort(n_levels))
76</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jeffersonfparil/GenomicBreeding.jl/blob/95e45ff7b84ffd98d4238d1241678add8e646bbd/src/models/tebv.jl#L150-L179">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicBreeding.writeJLD2-Tuple{GenomicBreeding.AbstractGB}" href="#GenomicBreeding.writeJLD2-Tuple{GenomicBreeding.AbstractGB}"><code>GenomicBreeding.writeJLD2</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">writeJLD2(A::Union{Genomes,Phenomes,Trials,SimulatedEffects}; fname::Union{Missing,String} = missing)::String</code></pre><p>Save core (<code>Genomes</code>, <code>Phenomes</code>, and <code>Trials</code>), simulation (<code>SimulatedEffects</code>), and model (<code>TEBV</code>) structs as JLD2,  a heirarchical data format version 5 (HDF5) - compatible format. Note that the extension name should be &#39;.jld2&#39;.</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; genomes = simulategenomes(n=2, verbose=false);

julia&gt; writeJLD2(genomes, fname=&quot;test_genomes.jld2&quot;)
&quot;test_genomes.jld2&quot;

julia&gt; load(&quot;test_genomes.jld2&quot;)[&quot;Genomes&quot;] == genomes
true

julia&gt; phenomes = Phenomes(n=2, t=2); phenomes.entries = [&quot;entry_1&quot;, &quot;entry_2&quot;]; phenomes.traits = [&quot;trait_1&quot;, &quot;trait_2&quot;];

julia&gt; writeJLD2(phenomes, fname=&quot;test_phenomes.jld2&quot;)
&quot;test_phenomes.jld2&quot;

julia&gt; load(&quot;test_phenomes.jld2&quot;)[&quot;Phenomes&quot;] == phenomes
true

julia&gt; trials = Trials(n=1, t=2); trials.entries = [&quot;entry_1&quot;];

julia&gt; writeJLD2(trials, fname=&quot;test_trials.jld2&quot;)
&quot;test_trials.jld2&quot;

julia&gt; load(&quot;test_trials.jld2&quot;)[&quot;Trials&quot;] == trials
true

julia&gt; simulated_effects = SimulatedEffects();

julia&gt; writeJLD2(simulated_effects, fname=&quot;test_simulated_effects.jld2&quot;)
&quot;test_simulated_effects.jld2&quot;

julia&gt; load(&quot;test_simulated_effects.jld2&quot;)[&quot;SimulatedEffects&quot;] == simulated_effects
true

julia&gt; trials, _simulated_effects = simulatetrials(genomes = simulategenomes(n=10, verbose=false), n_years=1, n_seasons=1, n_harvests=1, n_sites=1, n_replications=10, verbose=false);

julia&gt; tebv = analyse(trials, max_levels=50, verbose=false);

julia&gt; writeJLD2(tebv, fname=&quot;test_tebv.jld2&quot;)
&quot;test_tebv.jld2&quot;

julia&gt; load(&quot;test_tebv.jld2&quot;)[&quot;TEBV&quot;] == tebv
true</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jeffersonfparil/GenomicBreeding.jl/blob/95e45ff7b84ffd98d4238d1241678add8e646bbd/src/io/writer.jl#L1-L52">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicBreeding.writedelimited-Tuple{Genomes}" href="#GenomicBreeding.writedelimited-Tuple{Genomes}"><code>GenomicBreeding.writedelimited</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">writedelimited(genomes::Genomes, sep::String = &quot;	&quot;, fname::Union{Missing,String} = missing)::String</code></pre><p>Save <code>Genomes</code> struct as a string-delimited (default=<code></code>) file. Each row corresponds to a locus-allele combination. The first 4 columns correspond to the chromosome, position, all alleles in the locus (delimited by <code>|</code>), and the specific allele. The subsequency columns refer to the samples, pools, entries or genotypes.</p><p><strong>Notes:</strong></p><ul><li>Extension name should be &#39;.tsv&#39;, &#39;.csv&#39;, or &#39;.txt&#39;.</li><li>Header lines and comments are prefixed by &#39;#&#39;.</li><li>There are 2 header lines prefixed by &#39;#&#39;, e.g.:<ul><li>header line 1: &quot;chrom,pos,all<em>alleles,allele,entry</em>1,entry_2&quot;</li><li>header line 2: &quot;chrom,pos,all<em>alleles,allele,population</em>1,population_1&quot;</li></ul></li></ul><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; genomes = simulategenomes(n=2, verbose=false);

julia&gt; writedelimited(genomes, fname=&quot;test_genomes.tsv&quot;)
&quot;test_genomes.tsv&quot;</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jeffersonfparil/GenomicBreeding.jl/blob/95e45ff7b84ffd98d4238d1241678add8e646bbd/src/io/writer.jl#L79-L101">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicBreeding.writedelimited-Tuple{Phenomes}" href="#GenomicBreeding.writedelimited-Tuple{Phenomes}"><code>GenomicBreeding.writedelimited</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">writedelimited(phenomes::Phenomes, sep::String = &quot;	&quot;, fname::Union{Missing,String} = missing)::String</code></pre><p>Save <code>Phenomes</code> struct as a string-delimited (default=<code></code>) file.  Each row corresponds to a samples, pools, entries or genotypes. The first 2 columns correspond to the entry and population names. The subsequency columns refer to the traits containing the phenotype values of each entry. Note that the extension name should be &#39;.tsv&#39;, &#39;.csv&#39;, or &#39;.txt&#39;.</p><p><strong>Notes:</strong></p><ul><li>Extension name should be &#39;.tsv&#39;, &#39;.csv&#39;, or &#39;.txt&#39;.</li><li>Header line and comments are prefixed by &#39;#&#39;.</li><li>There is 1 header line prefixed by &#39;#&#39;, e.g.: &quot;entry,population,trait<em>1,trait</em>2,trait_3&quot;</li></ul><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; phenomes = Phenomes(n=2, t=2); phenomes.entries = [&quot;entry_1&quot;, &quot;entry_2&quot;]; phenomes.traits = [&quot;trait_1&quot;, &quot;trait_2&quot;];

julia&gt; writedelimited(phenomes, fname=&quot;test_phenomes.tsv&quot;)
&quot;test_phenomes.tsv&quot;</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jeffersonfparil/GenomicBreeding.jl/blob/95e45ff7b84ffd98d4238d1241678add8e646bbd/src/io/writer.jl#L154-L175">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicBreeding.writedelimited-Tuple{Trials}" href="#GenomicBreeding.writedelimited-Tuple{Trials}"><code>GenomicBreeding.writedelimited</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">writedelimited(trials::Trials, sep::String = &quot;	&quot;, fname::Union{Missing,String} = missing)::String</code></pre><p>Save <code>Trials</code> struct as a string-delimited (default=<code></code>) file.  Each row corresponds to a samples, pools, entries or genotypes. The first 10 columns correspond to:</p><ol><li>years</li><li>seasons</li><li>harvests</li><li>sites</li><li>entries</li><li>populations</li><li>replications</li><li>blocks</li><li>rows</li><li>cols </li></ol><p>The subsequency columns refer to the traits containing the phenotype values.</p><p><strong>Notes:</strong></p><ul><li>Extension name should be &#39;.tsv&#39;, &#39;.csv&#39;, or &#39;.txt&#39;.</li><li>Header line and comments are prefixed by &#39;#&#39;.</li><li>There is 1 header line prefixed by &#39;#&#39;, e.g.: &quot;years,seasons,harvests, ..., trait<em>1,tratit</em>2,trait_3&quot;</li></ul><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; trials = Trials(n=1, t=2); trials.years = [&quot;year_1&quot;]; trials.seasons = [&quot;season_1&quot;]; trials.harvests = [&quot;harvest_1&quot;]; trials.sites = [&quot;site_1&quot;]; trials.entries = [&quot;entry_1&quot;]; trials.populations = [&quot;population_1&quot;]; trials.replications = [&quot;replication_1&quot;]; trials.blocks = [&quot;block_1&quot;]; trials.rows = [&quot;row_1&quot;]; trials.cols = [&quot;col_1&quot;]; trials.traits = [&quot;trait_1&quot;, &quot;trait_2&quot;];

julia&gt; writedelimited(trials, fname=&quot;test_trials.tsv&quot;)
&quot;test_trials.tsv&quot;</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jeffersonfparil/GenomicBreeding.jl/blob/95e45ff7b84ffd98d4238d1241678add8e646bbd/src/io/writer.jl#L226-L258">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicBreeding.@string2formula-Tuple{Any}" href="#GenomicBreeding.@string2formula-Tuple{Any}"><code>GenomicBreeding.@string2formula</code></a> — <span class="docstring-category">Macro</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">string2formula(x)</code></pre><p>Macro to <code>Meta.parse</code> a string into a formula.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jeffersonfparil/GenomicBreeding.jl/blob/95e45ff7b84ffd98d4238d1241678add8e646bbd/src/models/tebv.jl#L140-L144">source</a></section></article></article><nav class="docs-footer"><a class="docs-footer-nextpage" href="simulations/">Simulation »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="auto">Automatic (OS)</option><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option><option value="catppuccin-latte">catppuccin-latte</option><option value="catppuccin-frappe">catppuccin-frappe</option><option value="catppuccin-macchiato">catppuccin-macchiato</option><option value="catppuccin-mocha">catppuccin-mocha</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.8.0 on <span class="colophon-date" title="Friday 20 December 2024 06:37">Friday 20 December 2024</span>. Using Julia version 1.11.2.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
