<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Home · GenomicBreeding.jl</title><meta name="title" content="Home · GenomicBreeding.jl"/><meta property="og:title" content="Home · GenomicBreeding.jl"/><meta property="twitter:title" content="Home · GenomicBreeding.jl"/><meta name="description" content="Documentation for GenomicBreeding.jl."/><meta property="og:description" content="Documentation for GenomicBreeding.jl."/><meta property="twitter:description" content="Documentation for GenomicBreeding.jl."/><meta property="og:url" content="https://jeffersonfparil.github.io/GenomicBreeding.jl/"/><meta property="twitter:url" content="https://jeffersonfparil.github.io/GenomicBreeding.jl/"/><link rel="canonical" href="https://jeffersonfparil.github.io/GenomicBreeding.jl/"/><script data-outdated-warner src="assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="assets/documenter.js"></script><script src="search_index.js"></script><script src="siteinfo.js"></script><script src="../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/catppuccin-mocha.css" data-theme-name="catppuccin-mocha"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/catppuccin-macchiato.css" data-theme-name="catppuccin-macchiato"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/catppuccin-frappe.css" data-theme-name="catppuccin-frappe"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/catppuccin-latte.css" data-theme-name="catppuccin-latte"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href>GenomicBreeding.jl</a></span></div><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li class="is-active"><a class="tocitem" href>Home</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>Home</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Home</a></li></ul></nav><div class="docs-right"><a class="docs-navbar-link" href="https://github.com/jeffersonfparil/GenomicBreeding.jl" title="View the repository on GitHub"><span class="docs-icon fa-brands"></span><span class="docs-label is-hidden-touch">GitHub</span></a><a class="docs-navbar-link" href="https://github.com/jeffersonfparil/GenomicBreeding.jl/blob/main/docs/src/index.md" title="Edit source on GitHub"><span class="docs-icon fa-solid"></span></a><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><h1 id="GenomicBreeding"><a class="docs-heading-anchor" href="#GenomicBreeding">GenomicBreeding</a><a id="GenomicBreeding-1"></a><a class="docs-heading-anchor-permalink" href="#GenomicBreeding" title="Permalink"></a></h1><p>Documentation for <a href="https://github.com/jeffersonfparil/GenomicBreeding.jl">GenomicBreeding</a>.</p><ul><li><a href="#GenomicBreeding.Genomes"><code>GenomicBreeding.Genomes</code></a></li><li><a href="#GenomicBreeding.Phenomes"><code>GenomicBreeding.Phenomes</code></a></li><li><a href="#GenomicBreeding.Trials"><code>GenomicBreeding.Trials</code></a></li><li><a href="#GenomicBreeding.checkdims-Tuple{Trials}"><code>GenomicBreeding.checkdims</code></a></li><li><a href="#GenomicBreeding.checkdims-Tuple{Genomes}"><code>GenomicBreeding.checkdims</code></a></li><li><a href="#GenomicBreeding.checkdims-Tuple{Phenomes}"><code>GenomicBreeding.checkdims</code></a></li><li><a href="#GenomicBreeding.dimensions-Tuple{Genomes}"><code>GenomicBreeding.dimensions</code></a></li><li><a href="#GenomicBreeding.simulateeffects-Tuple{}"><code>GenomicBreeding.simulateeffects</code></a></li><li><a href="#GenomicBreeding.simulategenomes-Tuple{}"><code>GenomicBreeding.simulategenomes</code></a></li><li><a href="#GenomicBreeding.simulategenomiceffects-Tuple{}"><code>GenomicBreeding.simulategenomiceffects</code></a></li><li><a href="#GenomicBreeding.simulatetrials-Tuple{}"><code>GenomicBreeding.simulatetrials</code></a></li></ul><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicBreeding.Genomes" href="#GenomicBreeding.Genomes"><code>GenomicBreeding.Genomes</code></a> — <span class="docstring-category">Type</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><p><strong>Genomes struct</strong></p><p>Containes unique entries and loci_alleles where allele frequencies can have missing values</p><p><strong>Constructor</strong></p><pre><code class="language-julia hljs">Genomes(; n::Int = 1, p::Int = 2)</code></pre><p><strong>Fields</strong></p><ul><li><code>entries</code>: names of the <code>n</code> entries or samples</li><li><code>populations</code>: name/s of the population/s each entries or samples belong to</li><li><code>loci_alleles</code>: names of the <code>p</code> loci-alleles combinations (<code>p</code> = <code>l</code> loci x <code>a-1</code> alleles) including the chromsome or scaffold name, position, all alleles, and current allele separated by tabs</li><li><code>allele_frequencies</code>: <code>n x p</code> matrix of allele frequencies between 0 and 1 which can have missing values</li><li><code>mask</code>: <code>n x p</code> matrix of boolean mask for selective analyses and slicing</li></ul><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; genomes = Genomes(n=2, p=2)
Genomes([#undef, #undef], [#undef, #undef], [#undef, #undef], Union{Missing, Real}[#undef #undef; #undef #undef], Bool[0 0; 0 0])

julia&gt; fieldnames(Genomes)
(:entries, :populations, :loci_alleles, :allele_frequencies, :mask)

julia&gt; genomes.entries = [&quot;entry_1&quot;, &quot;entry_2&quot;];

julia&gt; genomes.populations = [&quot;pop_1&quot;, &quot;pop_1&quot;];

julia&gt; genomes.loci_alleles = [&quot;chr1_12345_A|T_A&quot;, &quot;chr2_678910_C|D_D&quot;];

julia&gt; genomes.allele_frequencies = [0.5 0.25; 0.9 missing];

julia&gt; genomes.mask = [true true; true false];

julia&gt; genomes
Genomes([&quot;entry_1&quot;, &quot;entry_2&quot;], [&quot;pop_1&quot;, &quot;pop_1&quot;], [&quot;chr1_12345_A|T_A&quot;, &quot;chr2_678910_C|D_D&quot;], Union{Missing, Real}[0.5 0.25; 0.9 missing], Bool[1 1; 1 0])</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jeffersonfparil/GenomicBreeding.jl/blob/dc7123bc7a5f0220c872fe6fb4173986d350af52/src/core/genomes.jl#L1-L40">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicBreeding.Phenomes" href="#GenomicBreeding.Phenomes"><code>GenomicBreeding.Phenomes</code></a> — <span class="docstring-category">Type</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><p><strong>Phenomes struct</strong></p><p>Constains unique entries and traits where phenotype data can have missing values</p><p><strong>Constructor</strong></p><pre><code class="language-julia hljs">Phenomes(; n::Int = 1, t::Int = 2)</code></pre><p><strong>Fields</strong></p><ul><li><code>entries</code>: names of the <code>n</code> entries or samples</li><li><code>populations</code>: name/s of the population/s each entries or samples belong to</li><li><code>traits</code>: names of the <code>t</code> traits</li><li><code>phenotypes</code>: <code>n x t</code> matrix of numeric (<code>R</code>) phenotype data which can have missing values</li><li><code>mask</code>: <code>n x t</code> matrix of boolean mask for selective analyses and slicing</li></ul><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; phenomes = Phenomes(n=2, t=2)
Phenomes([#undef, #undef], [#undef, #undef], [#undef, #undef], Union{Missing, Real}[#undef #undef; #undef #undef], Bool[0 0; 0 0])

julia&gt; fieldnames(Phenomes)
(:entries, :populations, :traits, :phenotypes, :mask)

julia&gt; phenomes.entries = [&quot;entry_1&quot;, &quot;entry_2&quot;];

julia&gt; phenomes.populations = [&quot;pop_A&quot;, &quot;pop_B&quot;];

julia&gt; phenomes.traits = [&quot;height&quot;, &quot;yield&quot;];

julia&gt; phenomes.phenotypes = [200.0 2.5; 150.0 missing];

julia&gt; phenomes.mask = [true true; true false];

julia&gt; phenomes
Phenomes([&quot;entry_1&quot;, &quot;entry_2&quot;], [&quot;pop_A&quot;, &quot;pop_B&quot;], [&quot;height&quot;, &quot;yield&quot;], Union{Missing, Real}[200.0 2.5; 150.0 missing], Bool[1 1; 1 0])</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jeffersonfparil/GenomicBreeding.jl/blob/dc7123bc7a5f0220c872fe6fb4173986d350af52/src/core/phenomes.jl#L1-L40">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicBreeding.Trials" href="#GenomicBreeding.Trials"><code>GenomicBreeding.Trials</code></a> — <span class="docstring-category">Type</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><p><strong>Trials struct</strong></p><p>Contains phenotype data across years, seasons, harvest, sites, populations, replications, blocks, rows, and columns</p><p><strong>Constructor</strong></p><pre><code class="language-julia hljs">Trials(; n::Int = 2, p::Int = 2)</code></pre><p><strong>Fields</strong></p><ul><li><code>phenotypes</code>: <code>n x t</code> matrix of numeric phenotype data which can have missing values</li><li><code>traits</code>: names of the traits <code>t</code> traits</li><li><code>years</code>: names of the years corresponding to each row in the phenotype matrix</li><li><code>seasons</code>: names of the seasons corresponding to each row in the phenotype matrix</li><li><code>harvests</code>: names of the harvests corresponding to each row in the phenotype matrix</li><li><code>sites</code>: names of the sites corresponding to each row in the phenotype matrix</li><li><code>replications</code>: names of the replications corresponding to each row in the phenotype matrix</li><li><code>blocks</code>: names of the blocks corresponding to each row in the phenotype matrix</li><li><code>rows</code>: names of the rows corresponding to each row in the phenotype matrix</li><li><code>cols</code>: names of the cols corresponding to each row in the phenotype matrix</li><li><code>entries</code>: names of the entries corresponding to each row in the phenotype matrix</li><li><code>populations</code>: names of the populations corresponding to each row in the phenotype matrix</li></ul><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; trials = Trials(n=1, t=2)
Trials(Union{Missing, Real}[#undef #undef], [#undef, #undef], [#undef], [#undef], [#undef], [#undef], [#undef], [#undef], [#undef], [#undef], [#undef], [#undef])

julia&gt; fieldnames(Trials)
(:phenotypes, :traits, :years, :seasons, :harvests, :sites, :replications, :blocks, :rows, :cols, :entries, :populations)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jeffersonfparil/GenomicBreeding.jl/blob/dc7123bc7a5f0220c872fe6fb4173986d350af52/src/core/trials.jl#L1-L34">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicBreeding.checkdims-Tuple{Genomes}" href="#GenomicBreeding.checkdims-Tuple{Genomes}"><code>GenomicBreeding.checkdims</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">checkdims(genomes::Genomes)::Bool</code></pre><p>Check dimension compatibility of the fields of the Genomes struct</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; genomes = Genomes(n=2,p=4);

julia&gt; checkdims(genomes)
true

julia&gt; genomes.entries = [&quot;beaking_change&quot;];

julia&gt; checkdims(genomes)
false</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jeffersonfparil/GenomicBreeding.jl/blob/dc7123bc7a5f0220c872fe6fb4173986d350af52/src/core/genomes.jl#L58-L75">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicBreeding.checkdims-Tuple{Phenomes}" href="#GenomicBreeding.checkdims-Tuple{Phenomes}"><code>GenomicBreeding.checkdims</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">checkdims(y::Phenomes)::Bool</code></pre><p>Check dimension compatibility of the fields of the Phenomes struct</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; y = Phenomes(n=2, t=2);

julia&gt; checkdims(y)
true

julia&gt; y.populations = [&quot;beaking_change&quot;];

julia&gt; checkdims(y)
false</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jeffersonfparil/GenomicBreeding.jl/blob/dc7123bc7a5f0220c872fe6fb4173986d350af52/src/core/phenomes.jl#L58-L75">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicBreeding.checkdims-Tuple{Trials}" href="#GenomicBreeding.checkdims-Tuple{Trials}"><code>GenomicBreeding.checkdims</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">checkdims(trials::Trials)::Bool</code></pre><p>Check dimension compatibility of the fields of the Trials struct</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; trials = Trials(n=1, t=2);

julia&gt; trials.entries = [&quot;entry_1&quot;];

julia&gt; checkdims(trials)
true

julia&gt; trials.entries = [&quot;entering_2_entries&quot;, &quot;instead_of_just_1&quot;];

julia&gt; checkdims(trials)
false</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jeffersonfparil/GenomicBreeding.jl/blob/dc7123bc7a5f0220c872fe6fb4173986d350af52/src/core/trials.jl#L66-L85">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicBreeding.dimensions-Tuple{Genomes}" href="#GenomicBreeding.dimensions-Tuple{Genomes}"><code>GenomicBreeding.dimensions</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">dimensions(genomes::Genomes)::Tuple{Int, Int, Int}</code></pre><p>Count the number of entries, populations, loci, and maximum number of alleles per locus in the Genomes struct</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; genomes = simulategenomes(n=100, l=1_000, n_alleles=4, verbose=false);

julia&gt; dimensions(genomes)
(100, 100, 3000, 1000, 4)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jeffersonfparil/GenomicBreeding.jl/blob/dc7123bc7a5f0220c872fe6fb4173986d350af52/src/core/genomes.jl#L94-L106">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicBreeding.simulateeffects-Tuple{}" href="#GenomicBreeding.simulateeffects-Tuple{}"><code>GenomicBreeding.simulateeffects</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><p><strong>Simulate effects</strong></p><p>Sample <code>p</code> x <code>q</code> effects from a multivariate normal distribution with <code>μ~Exp(λ)</code> and <code>Σ=μμ&#39;</code></p><p><strong>Arguments</strong></p><ul><li><code>p</code>: number of correlated effects to simulate (default = 2)</li><li><code>q</code>: number times to simulate the correlated effects from the same distribution (default = 1)</li><li><code>λ</code>: parameter of the exponential distritbution from which the means will be sampled from (default = 1.00)</li><li><code>seed</code>: randomisation seed (default = 42)</li></ul><p><strong>Output</strong></p><ul><li><code>p</code> x <code>q</code> matrix of correlated effects</li></ul><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; θ::Array{Real,2} = simulateeffects();

julia&gt; sum(abs.(θ - [-0.0886501800782904; -0.596478483888422])) &lt; 0.00001
true</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jeffersonfparil/GenomicBreeding.jl/blob/dc7123bc7a5f0220c872fe6fb4173986d350af52/src/simulation/simulate_trials.jl#L1-L23">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicBreeding.simulategenomes-Tuple{}" href="#GenomicBreeding.simulategenomes-Tuple{}"><code>GenomicBreeding.simulategenomes</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><p><strong>Simulate genomes</strong></p><p><strong>Arguments</strong></p><ul><li><code>n</code>: number of entries (default = 100)</li><li><code>l</code>: number of loci (default = 10_000)</li><li><code>n_chroms</code>: number of chromosomes (default = 7)</li><li><code>n_alleles</code>: number of alleles per locus (default = 2)</li><li><code>max_pos</code>: total length of the genome in base-pairs (bp) (default = 135<em>000</em>000)</li><li><code>ld_corr_50perc_kb</code>: distance in bp at which linkage expressed as correlation between a pair of loci is at 50% (default = 100_000)</li><li><code>seed</code>: psuedo-random number generator seed for replicability (default = 42)</li><li><code>verbose</code>: Show progress bar and plot the linkage heatmap into an svg file? (default = true)</li></ul><p><strong>Output</strong></p><ul><li><code>Genomes</code></li></ul><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; genomes = simulategenomes(n=100, l=10_000, n_alleles=3, verbose=false);

julia&gt; length(genomes.entries)
100

julia&gt; length(genomes.populations)
100

julia&gt; length(genomes.loci_alleles)
20000

julia&gt; size(genomes.allele_frequencies)
(100, 20000)

julia&gt; rng::TaskLocalRNG = Random.seed!(123);

julia&gt; idx = StatsBase.sample(rng, range(1, 20_000, step=2), 250, replace = false, ordered = true);

julia&gt; correlations = StatsBase.cor(genomes.allele_frequencies[:, idx]);

julia&gt; correlations[10,10] == 1.00
true

julia&gt; correlations[10,10] &gt; correlations[10,250]
true</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jeffersonfparil/GenomicBreeding.jl/blob/dc7123bc7a5f0220c872fe6fb4173986d350af52/src/simulation/simulate_genomes.jl#L1-L45">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicBreeding.simulategenomiceffects-Tuple{}" href="#GenomicBreeding.simulategenomiceffects-Tuple{}"><code>GenomicBreeding.simulategenomiceffects</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><p><strong>Simulate genomic effects</strong></p><p>Simulate additive, dominance, and epistatic effects</p><p><strong>Arguments</strong></p><ul><li><code>genomes</code>: Genome struct includes the <code>n</code> entries x <code>p</code> loci-alleles combinations (<code>p</code> = <code>l</code> loci x <code>a-1</code> alleles)</li><li><code>f_additive</code>: proportion of the <code>l</code> loci with non-zero additive effects on the phenotype</li><li><code>f_dominance</code>: proportion of the <code>l*f_additive</code> additive effects loci with additional dominance effects</li><li><code>f_epistasis</code>: proportion of the <code>l*f_additive</code> additive effects loci with additional epistasis effects</li></ul><p><strong>Outputs</strong></p><ul><li><code>n</code> x <code>3</code> matrix of additive, dominance and epistasis effects per entry</li><li><code>p</code> x <code>3</code> matrix of additive, dominance and epistasis effects per locus-allele combination</li></ul><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; genomes::Genomes = simulategenomes(n=100, l=2_000, n_alleles=3, verbose=false);

julia&gt; G, B = simulategenomiceffects(genomes=genomes, f_additive=0.05, f_dominance=0.75, f_epistasis=0.25);

julia&gt; size.([G, B])
2-element Vector{Tuple{Int64, Int64}}:
 (100, 3)
 (4000, 3)

julia&gt; sum(B .!= 0.0, dims=1)
1×3 Matrix{Int64}:
 200  75  50</code></pre><p><strong>Details</strong></p><p>The additive, dominance, and epistasis allele effects share a common exponential distribution (<code>λ=1</code>) from which  the mean of the effects (<code>μ</code>) are sampled, and the covariance matrix is derived (<code>Σ = μ * μ&#39;</code>;  where if <code>det(Σ)≈0</code> then we iteratively add 1.00 to the diagonals until it becomes invertible or 10 iterations  finishes and throws an error). The non-additive or epistasis allele effects were simulated by multiplying the allele  frequencies of all possible unique pairs of epistasis alleles and their effects.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jeffersonfparil/GenomicBreeding.jl/blob/dc7123bc7a5f0220c872fe6fb4173986d350af52/src/simulation/simulate_trials.jl#L44-L81">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicBreeding.simulatetrials-Tuple{}" href="#GenomicBreeding.simulatetrials-Tuple{}"><code>GenomicBreeding.simulatetrials</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><p><strong>Simulate trials</strong></p><p><strong>Arguments</strong></p><ul><li><code>genomes</code>: Genome struct includes the <code>n</code> entries x <code>p</code> loci-alleles combinations (<code>p</code> = <code>l</code> loci x <code>a-1</code> alleles)</li><li><code>n_traits</code>: Number of traits (default = 3)</li><li><code>n_years</code>: Number of years (default = 2)</li><li><code>n_seasons</code>: Number of seasons (default = 4)</li><li><code>n_harvests</code>: Number of harvests (default = 2)</li><li><code>n_sites</code>: Number of sites (default = 4)</li><li><code>n_replications</code>: Number of replications (default = 2)</li><li><code>n_blocks</code>: Number of blocks across the entire field layout (default = missing)</li><li><code>n_rows</code>: Number of rows across the entire field layout (default = missing)</li><li><code>n_cols</code>: Number of columns across the entire field layout (default = missing)</li><li><code>proportion_of_variance</code>: <code>12</code> x <code>n_traits</code> numeric matrix of scaled/non-scaled proportion of variances allocated to   genetic and environmental effects (default = missing; values will be sampled from a uniform distribution  followed by a biased sample on the first row, i.e. additive effects row).  The rows correspond to the variance allocated to:<ol><li>additive genetic effects</li><li>dominance genetic effects</li><li>epistasis genetic effects</li><li>years effects</li><li>seasons effects</li><li>harvests effects</li><li>sites effects</li><li>replications effects</li><li>blocks effects</li><li>rows effects</li><li>cols effects</li><li>complex interaction effects</li></ol></li><li><code>seed</code>: Randomisation seed (default = 42)</li><li><code>sparsity</code>: Proportion of missing data (default = 0.0)</li><li><code>verbose</code>: Show trials simulation progress bar? (default = true)</li></ul><p><strong>Output</strong></p><ul><li><code>Trials</code></li></ul><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; genomes::Genomes = simulategenomes(n=100, l=2_000, n_alleles=3, verbose=false);

julia&gt; trials::Trials = simulatetrials(genomes=genomes, sparsity=0.25, verbose=false);

julia&gt; size(trials.phenotypes)
(12800, 3)

julia&gt; size(trials.traits)
(3,)

julia&gt; unique(trials.entries) == genomes.entries
true

julia&gt; unique(trials.populations) == unique(genomes.populations)
true

julia&gt; abs(mean(ismissing.(trials.phenotypes)) - 0.25) &lt; 0.00001
true</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jeffersonfparil/GenomicBreeding.jl/blob/dc7123bc7a5f0220c872fe6fb4173986d350af52/src/simulation/simulate_trials.jl#L194-L253">source</a></section></article></article><nav class="docs-footer"><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="auto">Automatic (OS)</option><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option><option value="catppuccin-latte">catppuccin-latte</option><option value="catppuccin-frappe">catppuccin-frappe</option><option value="catppuccin-macchiato">catppuccin-macchiato</option><option value="catppuccin-mocha">catppuccin-mocha</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.8.0 on <span class="colophon-date" title="Monday 9 December 2024 05:30">Monday 9 December 2024</span>. Using Julia version 1.11.2.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
