<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Home · GenomicBreeding.jl</title><meta name="title" content="Home · GenomicBreeding.jl"/><meta property="og:title" content="Home · GenomicBreeding.jl"/><meta property="twitter:title" content="Home · GenomicBreeding.jl"/><meta name="description" content="Documentation for GenomicBreeding.jl."/><meta property="og:description" content="Documentation for GenomicBreeding.jl."/><meta property="twitter:description" content="Documentation for GenomicBreeding.jl."/><meta property="og:url" content="https://GenomicBreeding.github.io/GenomicBreeding.jl/"/><meta property="twitter:url" content="https://GenomicBreeding.github.io/GenomicBreeding.jl/"/><link rel="canonical" href="https://GenomicBreeding.github.io/GenomicBreeding.jl/"/><script data-outdated-warner src="assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="assets/documenter.js"></script><script src="search_index.js"></script><script src="siteinfo.js"></script><script src="../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/catppuccin-mocha.css" data-theme-name="catppuccin-mocha"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/catppuccin-macchiato.css" data-theme-name="catppuccin-macchiato"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/catppuccin-frappe.css" data-theme-name="catppuccin-frappe"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/catppuccin-latte.css" data-theme-name="catppuccin-latte"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href>GenomicBreeding.jl</a></span></div><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li class="is-active"><a class="tocitem" href>Home</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>Home</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Home</a></li></ul></nav><div class="docs-right"><a class="docs-navbar-link" href="https://github.com/GenomicBreeding/GenomicBreeding.jl" title="View the repository on GitHub"><span class="docs-icon fa-brands"></span><span class="docs-label is-hidden-touch">GitHub</span></a><a class="docs-navbar-link" href="https://github.com/GenomicBreeding/GenomicBreeding.jl/blob/main/docs/src/index.md" title="Edit source on GitHub"><span class="docs-icon fa-solid"></span></a><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><h1 id="GenomicBreeding"><a class="docs-heading-anchor" href="#GenomicBreeding">GenomicBreeding</a><a id="GenomicBreeding-1"></a><a class="docs-heading-anchor-permalink" href="#GenomicBreeding" title="Permalink"></a></h1><p>Documentation for <a href="https://github.com/jeffersonfparil/GenomicBreeding.jl">GenomicBreeding</a>.</p><ul><li><a href="#GenomicBreeding.GBInput"><code>GenomicBreeding.GBInput</code></a></li><li><a href="#GenomicBreeding.assess-Tuple{GBInput}"><code>GenomicBreeding.assess</code></a></li><li><a href="#GenomicBreeding.extracteffects-Tuple{GBInput}"><code>GenomicBreeding.extracteffects</code></a></li><li><a href="#GenomicBreeding.load-Tuple{GBInput}"><code>GenomicBreeding.load</code></a></li><li><a href="#GenomicBreeding.submitslurmarrayjobs-Tuple{}"><code>GenomicBreeding.submitslurmarrayjobs</code></a></li></ul><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicBreeding.GBInput" href="#GenomicBreeding.GBInput"><code>GenomicBreeding.GBInput</code></a> — <span class="docstring-category">Type</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">mutable struct GBInput
    fname_geno::String
    fname_pheno::String
    bulk_cv::Bool
    populations::Union{Nothing, String, Vector{String}}
    traits::Union{Nothing, String, Vector{String}}
    models::Any
    n_folds::Int64
    n_replications::Int64
    maf::Float64
    mtv::Float64
    n_iter::Int64,
    n_burnin::Int64,
    fname_out_prefix::String,
    verbose::Bool
end</code></pre><p>Input struct (belongs to GBCore.AbstractGB type)</p><ul><li><p><code>fname_geno</code>: genotype file (see file format guide: <strong>TODO:</strong> {URL HERE})</p></li><li><p><code>fname_pheno</code>: phenotype file (see file format guide: <strong>TODO:</strong> {URL HERE})</p></li><li><p><code>bulk_cv</code>: perform cross-validation across all populations, i.e. disregard population grouping (Default = false)</p></li><li><p><code>populations</code>: include only these populations (Default = nothing which means include all populations)</p></li><li><p><code>traits</code>: include only these traits (Default = nothing which means include all traits)</p></li><li><p><code>models</code>: include these model functions (Default = [ridge, bayesa]; see models list: <strong>TODO:</strong> {URL HERE})</p></li><li><p><code>n_folds</code>: number of k partitions for k-fold cross-validation (Default = 5)</p></li><li><p><code>n_replications</code>: number of replications for repeated k-fold cross-validation (Default = 5)</p></li><li><p><code>keep_all</code>: keep all entries upon merging genomes and phenomes potentially resulting in sparsities in both structs? (Default = false)</p></li><li><p><code>maf</code>: minimum allele frequency (Default = 0.05)</p></li><li><p><code>mtv</code>: minimum trait variance (Default = 1e-7)</p></li><li><p><code>n_iter</code>: number of Bayesian model fitting MCMC/HMC iteration (Default = 1_500)</p></li><li><p><code>fname_out_prefix</code>: prefix of the output files which may include directory names (Default = &quot;&quot; which translates to <code>./GBOuput/output-&lt;yyyymmddHHMMSS&gt;-&lt;3_digit_random_number&gt;-</code>)</p></li><li><p><code>n_burnin</code>: number of initial Bayesian model fitting MCMC/HMC iterations to be excluded from the posterior distribution (Default = 500)</p></li><li><p><code>verbose</code>: show messages (Default = true)</p></li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GenomicBreeding.jl/blob/1889beab34cd95c1a99778ec16608e4ad3a153f4/src/io.jl#L1-L37">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicBreeding.assess-Tuple{GBInput}" href="#GenomicBreeding.assess-Tuple{GBInput}"><code>GenomicBreeding.assess</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">assess(input::GBInput)::Tuple{Vector{String},Vector{String}}</code></pre><p>Assess genomic prediction accuracy via replicated k-fold cross-validation. Outputs are saved as JLD2 (each containing a CV struct) and TXT (each line is a note on failed runs) files.</p><p><strong>Example</strong></p><pre><code class="language-julia-repl hljs">julia&gt; genomes = GBCore.simulategenomes(n=300, l=1_000, verbose=false); genomes.populations = StatsBase.sample(string.(&quot;pop_&quot;, 1:3), length(genomes.entries), replace=true);

julia&gt; proportion_of_variance = fill(0.0, 9, 3); proportion_of_variance[1, :] .= 1.00; # 100% variance on the additive genetic effects

julia&gt; trials, _ = GBCore.simulatetrials(genomes=genomes, n_years=1, n_seasons=1, n_harvests=1, n_sites=1, n_replications=1, proportion_of_variance=proportion_of_variance, verbose=false);

julia&gt; phenomes = extractphenomes(trials);

julia&gt; fname_geno = try writedelimited(genomes, fname=&quot;test-geno.tsv&quot;); catch; rm(&quot;test-geno.tsv&quot;); writedelimited(genomes, fname=&quot;test-geno.tsv&quot;); end;

julia&gt; fname_pheno = try writedelimited(phenomes, fname=&quot;test-pheno.tsv&quot;); catch; rm(&quot;test-pheno.tsv&quot;); writedelimited(phenomes, fname=&quot;test-pheno.tsv&quot;); end;

julia&gt; input = GBInput(fname_geno=fname_geno, fname_pheno=fname_pheno, populations=[&quot;pop_1&quot;, &quot;pop_3&quot;], traits=[&quot;trait_1&quot;], n_replications=2, n_folds=3, verbose=false);

julia&gt; fnames_cvs, fnames_notes = assess(input);

julia&gt; length(fnames_cvs) == 32, length(fnames_notes) == 0
(true, true)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GenomicBreeding.jl/blob/1889beab34cd95c1a99778ec16608e4ad3a153f4/src/genomic_prediction.jl#L1-L28">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicBreeding.extracteffects-Tuple{GBInput}" href="#GenomicBreeding.extracteffects-Tuple{GBInput}"><code>GenomicBreeding.extracteffects</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">extracteffects(input::GBInput)::Vector{String}</code></pre><p>Extract allele effects by fitting the models without cross-validation</p><p><strong>Example</strong></p><pre><code class="language-julia-repl hljs">julia&gt; genomes = GBCore.simulategenomes(n=300, l=1_000, verbose=false); genomes.populations = StatsBase.sample(string.(&quot;pop_&quot;, 1:3), length(genomes.entries), replace=true);

julia&gt; proportion_of_variance = fill(0.0, 9, 3); proportion_of_variance[1, :] .= 1.00; # 100% variance on the additive genetic effects

julia&gt; trials, _ = GBCore.simulatetrials(genomes=genomes, n_years=1, n_seasons=1, n_harvests=1, n_sites=1, n_replications=1, proportion_of_variance=proportion_of_variance, verbose=false);

julia&gt; phenomes = extractphenomes(trials);

julia&gt; fname_geno = try writedelimited(genomes, fname=&quot;test-geno.tsv&quot;); catch; rm(&quot;test-geno.tsv&quot;); writedelimited(genomes, fname=&quot;test-geno.tsv&quot;); end;

julia&gt; fname_pheno = try writedelimited(phenomes, fname=&quot;test-pheno.tsv&quot;); catch; rm(&quot;test-pheno.tsv&quot;); writedelimited(phenomes, fname=&quot;test-pheno.tsv&quot;); end;

julia&gt; input = GBInput(fname_geno=fname_geno, fname_pheno=fname_pheno, populations=[&quot;pop_1&quot;, &quot;pop_3&quot;], traits=[&quot;trait_1&quot;], n_replications=2, n_folds=3, verbose=false);

julia&gt; fname_effects = extracteffects(input);

julia&gt; length(fname_effects) == 6
true</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GenomicBreeding.jl/blob/1889beab34cd95c1a99778ec16608e4ad3a153f4/src/genomic_prediction.jl#L118-L144">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicBreeding.load-Tuple{GBInput}" href="#GenomicBreeding.load-Tuple{GBInput}"><code>GenomicBreeding.load</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">load(input::GBInput)::Tuple{Genomes, Phenomes}</code></pre><p>Load, merge and filter genotype and phenotype data</p><p><strong>Example</strong></p><pre><code class="language-julia-repl hljs">julia&gt; genomes = GBCore.simulategenomes(n=300, verbose=false); genomes.populations = StatsBase.sample(string.(&quot;pop_&quot;, 1:3), length(genomes.entries), replace=true);

julia&gt; trials, _ = GBCore.simulatetrials(genomes=genomes, n_years=1, n_seasons=1, n_harvests=1, n_sites=1, n_replications=1, verbose=false);

julia&gt; phenomes = extractphenomes(trials);

julia&gt; fname_geno = try writedelimited(genomes, fname=&quot;test-geno.tsv&quot;); catch; rm(&quot;test-geno.tsv&quot;); writedelimited(genomes, fname=&quot;test-geno.tsv&quot;); end;
    
julia&gt; fname_pheno = try writedelimited(phenomes, fname=&quot;test-pheno.tsv&quot;); catch; rm(&quot;test-pheno.tsv&quot;); writedelimited(phenomes, fname=&quot;test-pheno.tsv&quot;); end;

julia&gt; input = GBInput(fname_geno=fname_geno, fname_pheno=fname_pheno, populations=[&quot;pop_1&quot;, &quot;pop_3&quot;], traits=[&quot;trait_1&quot;], verbose=false);

julia&gt; genomes, phenomes = load(input);

julia&gt; length(unique(genomes.populations)) == length(unique(phenomes.populations)) == 2
true

julia&gt; length(phenomes.traits) == 1
true

julia&gt; rm.([fname_geno, fname_pheno]);</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GenomicBreeding.jl/blob/1889beab34cd95c1a99778ec16608e4ad3a153f4/src/io.jl#L96-L125">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomicBreeding.submitslurmarrayjobs-Tuple{}" href="#GenomicBreeding.submitslurmarrayjobs-Tuple{}"><code>GenomicBreeding.submitslurmarrayjobs</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">submitslurmarrayjobs(
    fname_geno::String,
    fname_pheno::String;
    models::Any = [ridge, bayesa],
    n_folds::Int64 = 5,
    n_replications::Int64 = 5,
    maf::Float64 = 0.05,
    mtv::Float64 = 1e-7,
    n_iter::Int64 = 1_500,
    n_burnin::Int64 = 500,
    fname_out_prefix::String = &quot;&quot;,
    verbose::Bool = true,
)::Vector{GBInput}</code></pre><p>TODO: submit Slurm array job</p><p><strong>Example</strong></p><pre><code class="language-julia-repl hljs">julia&gt; genomes = GBCore.simulategenomes(n=300, verbose=false); genomes.populations = StatsBase.sample(string.(&quot;pop_&quot;, 1:3), length(genomes.entries), replace=true);

julia&gt; trials, _ = GBCore.simulatetrials(genomes=genomes, n_years=1, n_seasons=1, n_harvests=1, n_sites=1, n_replications=1, verbose=false);

julia&gt; phenomes = extractphenomes(trials);

julia&gt; fname_geno = try writedelimited(genomes, fname=&quot;test-geno.tsv&quot;); catch; rm(&quot;test-geno.tsv&quot;); writedelimited(genomes, fname=&quot;test-geno.tsv&quot;); end;
    
julia&gt; fname_pheno = try writedelimited(phenomes, fname=&quot;test-pheno.tsv&quot;); catch; rm(&quot;test-pheno.tsv&quot;); writedelimited(phenomes, fname=&quot;test-pheno.tsv&quot;); end;

julia&gt; inputs = submitslurmarrayjobs(fname_geno=fname_geno, fname_pheno=fname_pheno, verbose=false);

julia&gt; length(inputs) == 24
true

julia&gt; rm.([fname_geno, fname_pheno]);</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GenomicBreeding.jl/blob/1889beab34cd95c1a99778ec16608e4ad3a153f4/src/io.jl#L357-L393">source</a></section></article></article><nav class="docs-footer"><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="auto">Automatic (OS)</option><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option><option value="catppuccin-latte">catppuccin-latte</option><option value="catppuccin-frappe">catppuccin-frappe</option><option value="catppuccin-macchiato">catppuccin-macchiato</option><option value="catppuccin-mocha">catppuccin-mocha</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.8.0 on <span class="colophon-date" title="Friday 7 February 2025 05:20">Friday 7 February 2025</span>. Using Julia version 1.11.3.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
