<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Home · GenomicBreeding.jl</title><meta name="title" content="Home · GenomicBreeding.jl"/><meta property="og:title" content="Home · GenomicBreeding.jl"/><meta property="twitter:title" content="Home · GenomicBreeding.jl"/><meta name="description" content="Documentation for GenomicBreeding.jl."/><meta property="og:description" content="Documentation for GenomicBreeding.jl."/><meta property="twitter:description" content="Documentation for GenomicBreeding.jl."/><meta property="og:url" content="https://GenomicBreeding.github.io/GenomicBreeding.jl/"/><meta property="twitter:url" content="https://GenomicBreeding.github.io/GenomicBreeding.jl/"/><link rel="canonical" href="https://GenomicBreeding.github.io/GenomicBreeding.jl/"/><script data-outdated-warner src="assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="assets/documenter.js"></script><script src="search_index.js"></script><script src="siteinfo.js"></script><script src="../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/catppuccin-mocha.css" data-theme-name="catppuccin-mocha"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/catppuccin-macchiato.css" data-theme-name="catppuccin-macchiato"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/catppuccin-frappe.css" data-theme-name="catppuccin-frappe"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/catppuccin-latte.css" data-theme-name="catppuccin-latte"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href>GenomicBreeding.jl</a></span></div><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li class="is-active"><a class="tocitem" href>Home</a><ul class="internal"><li><a class="tocitem" href="#Overview"><span>Overview</span></a></li><li><a class="tocitem" href="#Installation"><span>Installation</span></a></li><li><a class="tocitem" href="#Quickstart"><span>Quickstart</span></a></li><li><a class="tocitem" href="#License"><span>License</span></a></li><li><a class="tocitem" href="#main-index"><span>Index</span></a></li></ul></li><li><a class="tocitem" href="manual/">Manual</a></li><li><a class="tocitem" href="references/">Reference</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>Home</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Home</a></li></ul></nav><div class="docs-right"><a class="docs-navbar-link" href="https://github.com/GenomicBreeding/GenomicBreeding.jl" title="View the repository on GitHub"><span class="docs-icon fa-brands"></span><span class="docs-label is-hidden-touch">GitHub</span></a><a class="docs-navbar-link" href="https://github.com/GenomicBreeding/GenomicBreeding.jl/blob/main/docs/src/index.md" title="Edit source on GitHub"><span class="docs-icon fa-solid"></span></a><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><h1 id="GenomicBreeding.jl"><a class="docs-heading-anchor" href="#GenomicBreeding.jl">GenomicBreeding.jl</a><a id="GenomicBreeding.jl-1"></a><a class="docs-heading-anchor-permalink" href="#GenomicBreeding.jl" title="Permalink"></a></h1><p>Documentation for the <code>GenomicBreeding</code> module.</p><ul><li><a href="#GenomicBreeding.jl">GenomicBreeding.jl</a></li><li class="no-marker"><ul><li><a href="#Overview">Overview</a></li><li><a href="#Installation">Installation</a></li><li><a href="#Quickstart">Quickstart</a></li><li><a href="#License">License</a></li><li><a href="#main-index">Index</a></li></ul></li><li><a href="manual/#Manual">Manual</a></li><li class="no-marker"><ul><li><a href="manual/#Main">Main</a></li><li><a href="manual/#Core-Data-Structures">Core Data Structures</a></li><li><a href="manual/#General-Functions">General Functions</a></li><li><a href="manual/#Simulation-Functions">Simulation Functions</a></li><li><a href="manual/#Phenotype-Analysis-Functions">Phenotype Analysis Functions</a></li><li><a href="manual/#Input/Output-Functions-(GBIO)">Input/Output Functions (GBIO)</a></li><li><a href="manual/#Genomic-Prediction-Functions-(GBModels)">Genomic Prediction Functions (GBModels)</a></li><li><a href="manual/#Plotting-Functions-(GBPlots)">Plotting Functions (GBPlots)</a></li></ul></li><li><a href="references/#Methods-reference">Methods reference</a></li></ul><h2 id="Overview"><a class="docs-heading-anchor" href="#Overview">Overview</a><a id="Overview-1"></a><a class="docs-heading-anchor-permalink" href="#Overview" title="Permalink"></a></h2><p>The <code>GenomicBreeding</code> module provides a comprehensive suite of tools for genomic prediction, genome-wide association studies (GWAS), and data handling in genomic breeding. It integrates functionalities from <code>GBCore</code>, <code>GBIO</code>, <code>GBModels</code>, and <code>GBPlots</code> to offer efficient and scalable solutions for genetic data analysis and visualisation.</p><h2 id="Installation"><a class="docs-heading-anchor" href="#Installation">Installation</a><a id="Installation-1"></a><a class="docs-heading-anchor-permalink" href="#Installation" title="Permalink"></a></h2><p>We designed <a href="https://github.com/GenomicBreeding/GenomicBreeding.jl">GenomicBreeding.jl</a> to work on an HPC running Linux (the various components, i.e. <a href="https://github.com/GenomicBreeding/GBCore.jl">GBCore.jl</a>, <a href="https://github.com/GenomicBreeding/GBIO.jl">GBIO.jl</a>, <a href="https://github.com/GenomicBreeding/GBModels.jl">GBModels.jl</a>, and <a href="https://github.com/GenomicBreeding/GBPlots.jl">GBPlots.jl</a> work on a single Linux PC too).</p><p>Currently, we require that you install <a href="https://julialang.org/">Julia</a> on your home directory in you HPC cluster via:</p><pre><code class="language-shell hljs">curl -fsSL https://install.julialang.org | sh
type -a julia</code></pre><p>Currently, <a href="https://github.com/GenomicBreeding/GBModels.jl">GBModels.jl</a> is dependent on <a href="https://www.r-project.org/">R</a> and the package <a href="https://github.com/gdlc/BGLR-R">BGLR</a> for Bayes A, Bayes B and Bayes C models. Because of this we require that <a href="https://www.r-project.org/">R</a> and <a href="https://github.com/gdlc/BGLR-R">BGLR</a> be installed. To help with this, you may install all the requirements via <a href="https://www.anaconda.com/docs/getting-started/miniconda/main">Conda</a> using the environment file: <a href="https://github.com/GenomicBreeding/GenomicBreeding.jl/GenomicBreeding_conda.yml"><code>GenomicBreeding_conda.yml</code></a>. We aim to have a pure Julia implementation of Bayesian models using <a href="https://turinglang.org/">Turing.jl</a> in the near future (we just need to speed-up the models a bit).</p><p>Install the <a href="https://github.com/GenomicBreeding/GenomicBreeding.jl">GenomicBreeding.jl</a> library in Julia:</p><pre><code class="language-julia hljs">using Pkg
Pkg.add(&quot;https://github.com/GenomicBreeding/GenomicBreeding.jl&quot;)</code></pre><p>Feel free to install the <a href="https://github.com/GenomicBreeding">GenomicBreeding.jl components</a> as well as various other useful libraries:</p><pre><code class="language-julia hljs">using Pkg
GB_components = [
    &quot;https://github.com/GenomicBreeding/GBCore.jl&quot;,
    &quot;https://github.com/GenomicBreeding/GBIO.jl&quot;,
    &quot;https://github.com/GenomicBreeding/GBModels.jl&quot;,
    &quot;https://github.com/GenomicBreeding/GBPlots.jl&quot;,
]
for P in GB_components
    Pkg.add(url=P)
end
Pkg.add([&quot;StatsBase&quot;, &quot;MixedModels&quot;, &quot;MultivariateStats&quot;, &quot;UnicodePlots&quot;, &quot;ColorSchemes&quot;, &quot;CairoMakie&quot;])</code></pre><h2 id="Quickstart"><a class="docs-heading-anchor" href="#Quickstart">Quickstart</a><a id="Quickstart-1"></a><a class="docs-heading-anchor-permalink" href="#Quickstart" title="Permalink"></a></h2><ul><li><a href="#genomic-prediction">Genomic prediction</a></li><li><a href="#genome-wide-association-study">Genome-wide association study</a></li><li><a href="#genotype-data-filtering-and-imputation">Genotype data filtering and imputation</a></li><li><a href="#phenotype-data-analyses">Phenotype data analyses</a></li><li><a href="#cluster-analyses">Cluster analyses</a></li><li><a href="#plotting">Plotting</a></li><li><a href="#mating-simulations">Mating simulations</a></li></ul><h3 id="Genomic-prediction"><a class="docs-heading-anchor" href="#Genomic-prediction">Genomic prediction</a><a id="Genomic-prediction-1"></a><a class="docs-heading-anchor-permalink" href="#Genomic-prediction" title="Permalink"></a></h3><h4 id="Example-1:-using-simulated-data"><a class="docs-heading-anchor" href="#Example-1:-using-simulated-data">Example 1: using simulated data</a><a id="Example-1:-using-simulated-data-1"></a><a class="docs-heading-anchor-permalink" href="#Example-1:-using-simulated-data" title="Permalink"></a></h4><p>Here&#39;s a simple example using simulated data to perform replicated k-fold cross validation:</p><pre><code class="language-julia hljs"># It is always a good idea to keep all your packages updated
using Pkg; Pkg.update()
# Load GenomicBreeding
using GenomicBreeding
# Load plotting and GP model functions
import GenomicBreeding: plot, lasso, bayesa
# Simulate genotype and phenotype data
genomes = simulategenomes(n=300, l=1_000, verbose=true)
genomes.populations[1:100] .= &quot;pop1&quot;; genomes.populations[101:200] .= &quot;pop2&quot;; genomes.populations[201:300] .= &quot;pop3&quot; # simulate multiple populations
trials, _ = simulatetrials(genomes=genomes, n_years=1, n_seasons=1, n_harvests=1, n_sites=1, n_replications=1, verbose=true);
phenomes = extractphenomes(trials)
fname_geno = writedelimited(genomes, fname=&quot;test-geno.tsv&quot;)
fname_pheno = writedelimited(phenomes, fname=&quot;test-pheno.tsv&quot;)
# Input struct documentation
@doc GBInput
# Setup the input struct
input = GBInput(
    fname_geno=fname_geno, 
    fname_pheno=fname_pheno,
    anaysis=cv, # set analysis to use the `cv` function for replicated k-fold cross-validation
    models = [lasso, bayesa],
    n_folds=2, 
    n_replications=2, 
    SLURM_job_name=&quot;testGB&quot;,
    SLURM_account_name=&quot;dbiopast1&quot;,
    SLURM_cpus_per_task=2, 
    SLURM_mem_G=5, 
    SLURM_time_limit_dd_hhmmss=&quot;00-00:15:00&quot;,
    SLURM_max_array_jobs_running=10,
    verbose=true
)
# Preliminary look at the genotype and phenotype data
plot(input=input, format=&quot;png&quot;, plot_size=(700, 500))
# Documentation of the main user interface function (take note of the currently available analyses)
@doc submitslurmarrayjobs
# Submit the Slurm array jobs
# Note that here you will be prompted to enter YES to proceed.
outdir = submitslurmarrayjobs(input)
# Monitor the Slurm jobs
run(`sh -c &#39;squeue -u $USER&#39;`)
run(`sh -c &#39;ls -lhtr slurm-*_*.out&#39;`)
run(`sh -c &#39;cat slurm-*_*.out&#39;`)
run(`sh -c &#39;tail slurm-*_*.out&#39;`)
run(`sh -c &#39;grep -i &quot;err&quot; slurm-*_*.out&#39;`)
run(`sh -c &#39;grep -i &quot;err&quot; slurm-*_*.out | cut -d: -f1 | sort | uniq&#39;`)
readdir(outdir)
# Once the array jobs have finishes or at least a couple of jobs have finished, run below.
# Rerun as you often as wish to update the plots.
# You may exit Julia and just run the plotting function below after correctly defining input::GBInput above.
plot(input=input, format=&quot;png&quot;, plot_size=(700, 500), skip_genomes=true, skip_phenomes=true, overwrite=true)</code></pre><h4 id="Example-2:-using-empirical-data"><a class="docs-heading-anchor" href="#Example-2:-using-empirical-data">Example 2: using empirical data</a><a id="Example-2:-using-empirical-data-1"></a><a class="docs-heading-anchor-permalink" href="#Example-2:-using-empirical-data" title="Permalink"></a></h4><pre><code class="language-julia hljs"># TODO
# with GP per se on unphenotyped set</code></pre><h3 id="Genome-wide-association-study"><a class="docs-heading-anchor" href="#Genome-wide-association-study">Genome-wide association study</a><a id="Genome-wide-association-study-1"></a><a class="docs-heading-anchor-permalink" href="#Genome-wide-association-study" title="Permalink"></a></h3><pre><code class="language-julia hljs"># TODO</code></pre><h3 id="Genotype-data-filtering-and-imputation"><a class="docs-heading-anchor" href="#Genotype-data-filtering-and-imputation">Genotype data filtering and imputation</a><a id="Genotype-data-filtering-and-imputation-1"></a><a class="docs-heading-anchor-permalink" href="#Genotype-data-filtering-and-imputation" title="Permalink"></a></h3><pre><code class="language-julia hljs"># TODO</code></pre><h3 id="Phenotype-data-analyses"><a class="docs-heading-anchor" href="#Phenotype-data-analyses">Phenotype data analyses</a><a id="Phenotype-data-analyses-1"></a><a class="docs-heading-anchor-permalink" href="#Phenotype-data-analyses" title="Permalink"></a></h3><pre><code class="language-julia hljs"># TODO</code></pre><h3 id="Cluster-analyses"><a class="docs-heading-anchor" href="#Cluster-analyses">Cluster analyses</a><a id="Cluster-analyses-1"></a><a class="docs-heading-anchor-permalink" href="#Cluster-analyses" title="Permalink"></a></h3><pre><code class="language-julia hljs"># TODO</code></pre><h3 id="Plotting"><a class="docs-heading-anchor" href="#Plotting">Plotting</a><a id="Plotting-1"></a><a class="docs-heading-anchor-permalink" href="#Plotting" title="Permalink"></a></h3><pre><code class="language-julia hljs"># TODO</code></pre><h3 id="Mating-simulations"><a class="docs-heading-anchor" href="#Mating-simulations">Mating simulations</a><a id="Mating-simulations-1"></a><a class="docs-heading-anchor-permalink" href="#Mating-simulations" title="Permalink"></a></h3><pre><code class="language-julia hljs"># TODO</code></pre><h2 id="License"><a class="docs-heading-anchor" href="#License">License</a><a id="License-1"></a><a class="docs-heading-anchor-permalink" href="#License" title="Permalink"></a></h2><p>The <code>GenomicBreeding</code> module is licensed under the GPLv3 License. See the <a href="https://github.com/GenomicBreeding/GenomicBreeding.jl/blob/main/LICENSE.md">LICENSE.md</a> file for more details.</p><h2 id="main-index"><a class="docs-heading-anchor" href="#main-index">Index</a><a id="main-index-1"></a><a class="docs-heading-anchor-permalink" href="#main-index" title="Permalink"></a></h2><ul><li><a href="references/#GenomicBreeding.GBInput"><code>GenomicBreeding.GBInput</code></a></li><li><a href="references/#Base.:==-Tuple{GBInput, GBInput}"><code>Base.:==</code></a></li><li><a href="references/#Base.hash-Tuple{GBInput, UInt64}"><code>Base.hash</code></a></li><li><a href="references/#GBCore.checkdims-Tuple{GBInput}"><code>GBCore.checkdims</code></a></li><li><a href="references/#GBCore.clone-Tuple{GBInput}"><code>GBCore.clone</code></a></li><li><a href="references/#GenomicBreeding.checkinputs-Tuple{GBInput}"><code>GenomicBreeding.checkinputs</code></a></li><li><a href="references/#GenomicBreeding.cv-Tuple{GBInput}"><code>GenomicBreeding.cv</code></a></li><li><a href="references/#GenomicBreeding.fit-Tuple{GBInput}"><code>GenomicBreeding.fit</code></a></li><li><a href="references/#GenomicBreeding.gwas-Tuple{GBInput}"><code>GenomicBreeding.gwas</code></a></li><li><a href="references/#GenomicBreeding.loadcvs-Tuple{GBInput}"><code>GenomicBreeding.loadcvs</code></a></li><li><a href="references/#GenomicBreeding.loadfits-Tuple{GBInput}"><code>GenomicBreeding.loadfits</code></a></li><li><a href="references/#GenomicBreeding.loadgenomesphenomes-Tuple{GBInput}"><code>GenomicBreeding.loadgenomesphenomes</code></a></li><li><a href="references/#GenomicBreeding.plot-Tuple{}"><code>GenomicBreeding.plot</code></a></li><li><a href="references/#GenomicBreeding.predict-Tuple{GBInput}"><code>GenomicBreeding.predict</code></a></li><li><a href="references/#GenomicBreeding.prepareinputs-Tuple{GBInput}"><code>GenomicBreeding.prepareinputs</code></a></li><li><a href="references/#GenomicBreeding.prepareoutprefixandoutdir-Tuple{GBInput}"><code>GenomicBreeding.prepareoutprefixandoutdir</code></a></li><li><a href="references/#GenomicBreeding.submitslurmarrayjobs-Tuple{GBInput}"><code>GenomicBreeding.submitslurmarrayjobs</code></a></li></ul></article><nav class="docs-footer"><a class="docs-footer-nextpage" href="manual/">Manual »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="auto">Automatic (OS)</option><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option><option value="catppuccin-latte">catppuccin-latte</option><option value="catppuccin-frappe">catppuccin-frappe</option><option value="catppuccin-macchiato">catppuccin-macchiato</option><option value="catppuccin-mocha">catppuccin-mocha</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.9.0 on <span class="colophon-date" title="Friday 21 March 2025 05:55">Friday 21 March 2025</span>. Using Julia version 1.11.4.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
